{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Aging Biomarker Identification - Dataset generation\n",
    "\n",
    "\n",
    "## Overview of project\n",
    "\n",
    "This project seeks to identify novel genes that can serve as biomarkers for longevity and select candidates as therapeutic targets, using a strong and logical rationale based on biology and biologically relevant data. To that aim, we use the STRING database, which contains human protein-protein interaction data, and use a random walk with restart and identify gene connectivity with a set of established (i.e., known and proven) genes with impact on longevity, referred to as seed genes (sGenes), and extracted from the HAGR's GenAge (genage_human.csv). We will build a classifier that classifies genes into good or bad targets for longevity therapeutics: Longevity-Relevan Genes (LRGs) and Non-longevity Relevant Genes (NRGs). LRGs and NRGs are obtained from the targets of DrugAge database compounds, with LRGs having positive lifespan increases and NRGs having negative impacts. GenDR is used to augment the selection of LRGs, CellAge for sGenes, and \"Homo_sapiens.tsv\" for translating genes across species. Lastly, model evaluation and selection will yield an optimal model, which will be applied to non-training genes to identify novel LRGs.\n",
    "\n",
    "## Table of Contents\n",
    "\n",
    "1. [Longevity Genes](#cellage-data)\n",
    "2. [Gene Expression Data](#gene-expression-data)\n",
    "3. [Protein-Protein Interaction Data](#protein-protein-interaction-data)\n",
    "4. [Data Integration](#data-integration)\n",
    "5. [Data Preprocessing](#data-preprocessing)\n",
    "6. [Exploratory Data Analysis](#exploratory-data-analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Below we import all necessary libraries for running this notebook\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "from tqdm import tqdm\n",
    "pwd = os.getcwd()\n",
    "data_dir = os.path.join(pwd, '../data')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Longevity data\n",
    "\n",
    "We have collected cellular senescence data from the CellAge and GeneAge from the Human Ageing Genomic Resources (HAGR) databases, which provides information on human genes associated with aging and cellular senescence. \n",
    "\n",
    "From HAGR we also extract the DrugAge and GenDR, containing information of drugs that have anti-aging properties and genes affected by Dietary Restriction, respectively. This information is useful as it will provide a foundational definition for therapy-effect in the context of extending longevity.\n",
    "\n",
    "### Data Source\n",
    "\n",
    "- GenAge: [https://genomics.senescence.info/genes/human.html](https://genomics.senescence.info/genes/human.html)\n",
    "- CellAge: [https://genomics.senescence.info/download.html#cellage](https://genomics.senescence.info/download.html#cellage)\n",
    "- DrugAge: [https://genomics.senescence.info/drugs/](https://genomics.senescence.info/drugs/)\n",
    "- GenDR: [https://genomics.senescence.info/diet/](https://genomics.senescence.info/diet/)\n",
    "- BioMART (Ensembl): [http://www.ensembl.org/biomart](http://www.ensembl.org/biomart)\n",
    "\n",
    "### Dataset\n",
    "\n",
    "- genage_human.csv: List of manually curated genes identified through meta-analysis of aging human microarray data. \n",
    "- CellAge_genes.csv: List of genes experimentally identified as being associated with cellular senescence, along with experimental evidence, Gene Ontology terms, and orthologous genes in model organisms.\n",
    "- drugage.csv: List of compounds observed to have an impact on longevity in healthy organisms.\n",
    "- gendr_manipulations.csv: List of genes affected by dietary restriction.\n",
    "- Humo_sapiens.tsv: File containing human homologs for all genes in GeneDR\n",
    "- mart_export.txt: CSV file containing naming information for translating genes across datasets + GeneOntology terms for each gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GenAge ID</th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>entrez gene id</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>why</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GHR</td>\n",
       "      <td>growth hormone receptor</td>\n",
       "      <td>2690</td>\n",
       "      <td>GHR_HUMAN</td>\n",
       "      <td>mammal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>GHRH</td>\n",
       "      <td>growth hormone releasing hormone</td>\n",
       "      <td>2691</td>\n",
       "      <td>SLIB_HUMAN</td>\n",
       "      <td>mammal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>SHC1</td>\n",
       "      <td>SHC (Src homology 2 domain containing) transfo...</td>\n",
       "      <td>6464</td>\n",
       "      <td>SHC1_HUMAN</td>\n",
       "      <td>mammal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>POU1F1</td>\n",
       "      <td>POU class 1 homeobox 1</td>\n",
       "      <td>5449</td>\n",
       "      <td>PIT1_HUMAN</td>\n",
       "      <td>mammal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>PROP1</td>\n",
       "      <td>PROP paired-like homeobox 1</td>\n",
       "      <td>5626</td>\n",
       "      <td>PROP1_HUMAN</td>\n",
       "      <td>mammal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   GenAge ID  symbol                                               name  \\\n",
       "0          1     GHR                            growth hormone receptor   \n",
       "1          2    GHRH                   growth hormone releasing hormone   \n",
       "2          3    SHC1  SHC (Src homology 2 domain containing) transfo...   \n",
       "3          4  POU1F1                             POU class 1 homeobox 1   \n",
       "4          5   PROP1                        PROP paired-like homeobox 1   \n",
       "\n",
       "   entrez gene id      uniprot     why  \n",
       "0            2690    GHR_HUMAN  mammal  \n",
       "1            2691   SLIB_HUMAN  mammal  \n",
       "2            6464   SHC1_HUMAN  mammal  \n",
       "3            5449   PIT1_HUMAN  mammal  \n",
       "4            5626  PROP1_HUMAN  mammal  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entrez ID</th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Cancer Cell</th>\n",
       "      <th>Type of senescence</th>\n",
       "      <th>Senescence Effect</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22848</td>\n",
       "      <td>AAK1</td>\n",
       "      <td>AP2 associated kinase 1</td>\n",
       "      <td>No</td>\n",
       "      <td>Unclear</td>\n",
       "      <td>Induces</td>\n",
       "      <td>26583757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5243</td>\n",
       "      <td>ABCB1</td>\n",
       "      <td>ATP binding cassette subfamily B member 1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Stress-induced</td>\n",
       "      <td>Induces</td>\n",
       "      <td>10825123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>368</td>\n",
       "      <td>ABCC6</td>\n",
       "      <td>ATP binding cassette subfamily C member 6</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Unclear</td>\n",
       "      <td>Inhibits</td>\n",
       "      <td>28536638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>51225</td>\n",
       "      <td>ABI3</td>\n",
       "      <td>ABI family member 3</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Oncogene-induced</td>\n",
       "      <td>Induces</td>\n",
       "      <td>21223585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>25890</td>\n",
       "      <td>ABI3BP</td>\n",
       "      <td>ABI family member 3 binding protein</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Unclear</td>\n",
       "      <td>Induces</td>\n",
       "      <td>18559958</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Entrez ID Gene symbol                                  Gene name  \\\n",
       "0      22848        AAK1                    AP2 associated kinase 1   \n",
       "1       5243       ABCB1  ATP binding cassette subfamily B member 1   \n",
       "2        368       ABCC6  ATP binding cassette subfamily C member 6   \n",
       "3      51225        ABI3                        ABI family member 3   \n",
       "4      25890      ABI3BP        ABI family member 3 binding protein   \n",
       "\n",
       "  Cancer Cell Type of senescence Senescence Effect  Reference  \n",
       "0          No            Unclear           Induces   26583757  \n",
       "1         Yes     Stress-induced           Induces   10825123  \n",
       "2         Yes            Unclear          Inhibits   28536638  \n",
       "3         Yes   Oncogene-induced           Induces   21223585  \n",
       "4         Yes            Unclear           Induces   18559958  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genage_df = pd.read_csv(os.path.join(data_dir, 'HAGR/GenAge/genage_human.csv'))\n",
    "display(genage_df.head())\n",
    "cellage_df = pd.read_csv(os.path.join(\n",
    "    data_dir, 'HAGR/CellAge/cellage3.tsv'), sep='\\t', header=0)\n",
    "cellage_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>compound_name</th>\n",
       "      <th>cas_number</th>\n",
       "      <th>species</th>\n",
       "      <th>strain</th>\n",
       "      <th>dosage</th>\n",
       "      <th>avg_lifespan_change</th>\n",
       "      <th>max_lifespan_change</th>\n",
       "      <th>gender</th>\n",
       "      <th>significance</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ethanol</td>\n",
       "      <td>64-17-5</td>\n",
       "      <td>Drosophila mojavensis</td>\n",
       "      <td>A350</td>\n",
       "      <td>2%</td>\n",
       "      <td>57.65</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>O</td>\n",
       "      <td>13369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ethanol</td>\n",
       "      <td>64-17-5</td>\n",
       "      <td>Drosophila mojavensis</td>\n",
       "      <td>A350</td>\n",
       "      <td>2%</td>\n",
       "      <td>59.55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MALE</td>\n",
       "      <td>O</td>\n",
       "      <td>13369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ethanol</td>\n",
       "      <td>64-17-5</td>\n",
       "      <td>Drosophila mojavensis</td>\n",
       "      <td>A350</td>\n",
       "      <td>4%</td>\n",
       "      <td>28.42</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>O</td>\n",
       "      <td>13369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ethanol</td>\n",
       "      <td>64-17-5</td>\n",
       "      <td>Drosophila mojavensis</td>\n",
       "      <td>A350</td>\n",
       "      <td>4%</td>\n",
       "      <td>21.75</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MALE</td>\n",
       "      <td>O</td>\n",
       "      <td>13369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ethanol</td>\n",
       "      <td>64-17-5</td>\n",
       "      <td>Drosophila mojavensis</td>\n",
       "      <td>A420</td>\n",
       "      <td>2%</td>\n",
       "      <td>89.09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>O</td>\n",
       "      <td>13369</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  compound_name cas_number                species strain dosage  \\\n",
       "0       Ethanol    64-17-5  Drosophila mojavensis   A350     2%   \n",
       "1       Ethanol    64-17-5  Drosophila mojavensis   A350     2%   \n",
       "2       Ethanol    64-17-5  Drosophila mojavensis   A350     4%   \n",
       "3       Ethanol    64-17-5  Drosophila mojavensis   A350     4%   \n",
       "4       Ethanol    64-17-5  Drosophila mojavensis   A420     2%   \n",
       "\n",
       "   avg_lifespan_change  max_lifespan_change  gender significance  pubmed_id  \n",
       "0                57.65                  NaN  FEMALE            O      13369  \n",
       "1                59.55                  NaN    MALE            O      13369  \n",
       "2                28.42                  NaN  FEMALE            O      13369  \n",
       "3                21.75                  NaN    MALE            O      13369  \n",
       "4                89.09                  NaN  FEMALE            O      13369  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GenDR ID</th>\n",
       "      <th>gene symbol</th>\n",
       "      <th>species</th>\n",
       "      <th>entrez gene id</th>\n",
       "      <th>gene name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>SIR2</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>851520</td>\n",
       "      <td>Silent Information Regulator 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>CDC25</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>851019</td>\n",
       "      <td>Cell Division Cycle 25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>HAP4</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>853751</td>\n",
       "      <td>Heme Activator Protein 4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>PNC1</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>852846</td>\n",
       "      <td>Pyrazinamidase/NiCotinamidase 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>CYT1</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>854231</td>\n",
       "      <td>Cyt1p</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  GenDR ID gene symbol                   species entrez gene id  \\\n",
       "0        1        SIR2  Saccharomyces cerevisiae         851520   \n",
       "1        2       CDC25  Saccharomyces cerevisiae         851019   \n",
       "2        3        HAP4  Saccharomyces cerevisiae         853751   \n",
       "3        4        PNC1  Saccharomyces cerevisiae         852846   \n",
       "4        5        CYT1  Saccharomyces cerevisiae         854231   \n",
       "\n",
       "                         gene name  \n",
       "0   Silent Information Regulator 2  \n",
       "1           Cell Division Cycle 25  \n",
       "2         Heme Activator Protein 4  \n",
       "3  Pyrazinamidase/NiCotinamidase 1  \n",
       "4                            Cyt1p  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>species_name</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>homolog species</th>\n",
       "      <th>homolog entrez_id</th>\n",
       "      <th>homolog identifier</th>\n",
       "      <th>homolog symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>wwp-1</th>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>171647</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>11060</td>\n",
       "      <td>232</td>\n",
       "      <td>WWP2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wwp-1</th>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>171647</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>11059</td>\n",
       "      <td>232</td>\n",
       "      <td>WWP1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wwp-1</th>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>171647</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>83737</td>\n",
       "      <td>232</td>\n",
       "      <td>ITCH</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>let-363</th>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>172167</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2475</td>\n",
       "      <td>43</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vps-34</th>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>172280</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>5289</td>\n",
       "      <td>657</td>\n",
       "      <td>PIK3C3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    gene_symbol  species_name     entrez_id  homolog species  \\\n",
       "wwp-1    Caenorhabditis elegans        171647  Homo sapiens            11060   \n",
       "wwp-1    Caenorhabditis elegans        171647  Homo sapiens            11059   \n",
       "wwp-1    Caenorhabditis elegans        171647  Homo sapiens            83737   \n",
       "let-363  Caenorhabditis elegans        172167  Homo sapiens             2475   \n",
       "vps-34   Caenorhabditis elegans        172280  Homo sapiens             5289   \n",
       "\n",
       "         homolog entrez_id homolog identifier  homolog symbol  \n",
       "wwp-1                  232               WWP2             NaN  \n",
       "wwp-1                  232               WWP1             NaN  \n",
       "wwp-1                  232               ITCH             NaN  \n",
       "let-363                 43               MTOR             NaN  \n",
       "vps-34                 657             PIK3C3             NaN  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drugage_df = pd.read_csv(os.path.join(data_dir, 'HAGR/DrugAge/drugage.csv')\n",
    "                         ).drop(columns=[f'Unnamed: {i}' for i in range(10, 12)])\n",
    "display(drugage_df.head())\n",
    "gendr_df = pd.read_csv(os.path.join(data_dir, 'HAGR/GenDR/gendr_manipulations.csv')\n",
    "                       ).drop(columns=[f'Unnamed: {i}' for i in range(5, 286)])\n",
    "display(gendr_df.head())\n",
    "gendr_homologs_df = pd.read_csv(os.path.join(\n",
    "    data_dir, 'HAGR/GenDR/Homo_sapiens.tsv'), sep='\\t')\n",
    "gendr_homologs_df.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene Expression Data\n",
    "\n",
    "We have collected publicly available gene expression datasets from the Gene Expression Omnibus (GEO) and The Cancer Genome Atlas (TCGA). These datasets contain aging-related gene expression profiles for various tissue types and organisms.\n",
    "\n",
    "### Data Sources\n",
    "\n",
    "- GEO: [https://www.ncbi.nlm.nih.gov/geo/](https://www.ncbi.nlm.nih.gov/geo/)\n",
    "- TCGA: [https://portal.gdc.cancer.gov/](https://portal.gdc.cancer.gov/)\n",
    "\n",
    "### Datasets\n",
    "\n",
    "- GSEXXXX: Description of the dataset, species, tissue, number of samples, etc.\n",
    "- GSEYYYY: Description of the dataset, species, tissue, number of samples, etc.\n",
    "- ...\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Protein-Protein Interaction Data\n",
    "\n",
    "We have collected protein-protein interaction (PPI) data from STRING and BioGRID to incorporate network interactions as additional features for each gene.\n",
    "\n",
    "### Data Sources\n",
    "\n",
    "- STRING: [https://string-db.org/](https://string-db.org/)\n",
    "- BioGRID: [https://thebiogrid.org/](https://thebiogrid.org/)\n",
    "\n",
    "### Versions\n",
    "\n",
    "- STRING: Version X.X\n",
    "- BioGRID: Version Y.Y\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>neighborhood_transferred</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>homology</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>coexpression_transferred</th>\n",
       "      <th>experiments</th>\n",
       "      <th>experiments_transferred</th>\n",
       "      <th>database</th>\n",
       "      <th>database_transferred</th>\n",
       "      <th>textmining</th>\n",
       "      <th>textmining_transferred</th>\n",
       "      <th>combined_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000379496</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>54</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>103</td>\n",
       "      <td>85</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000314067</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>180</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000263116</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "      <td>0</td>\n",
       "      <td>152</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>101</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000361263</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>161</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>47</td>\n",
       "      <td>58</td>\n",
       "      <td>181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000409666</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>63</td>\n",
       "      <td>0</td>\n",
       "      <td>213</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>270</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               protein1              protein2  neighborhood  \\\n",
       "0  9606.ENSP00000000233  9606.ENSP00000379496             0   \n",
       "1  9606.ENSP00000000233  9606.ENSP00000314067             0   \n",
       "2  9606.ENSP00000000233  9606.ENSP00000263116             0   \n",
       "3  9606.ENSP00000000233  9606.ENSP00000361263             0   \n",
       "4  9606.ENSP00000000233  9606.ENSP00000409666             0   \n",
       "\n",
       "   neighborhood_transferred  fusion  cooccurence  homology  coexpression  \\\n",
       "0                         0       0            0         0             0   \n",
       "1                         0       0            0         0             0   \n",
       "2                         0       0            0         0             0   \n",
       "3                         0       0            0         0             0   \n",
       "4                         0       0            0         0            60   \n",
       "\n",
       "   coexpression_transferred  experiments  experiments_transferred  database  \\\n",
       "0                        54            0                        0         0   \n",
       "1                         0            0                      180         0   \n",
       "2                        62            0                      152         0   \n",
       "3                         0            0                      161         0   \n",
       "4                        63            0                      213         0   \n",
       "\n",
       "   database_transferred  textmining  textmining_transferred  combined_score  \n",
       "0                     0         103                      85             155  \n",
       "1                     0           0                      61             197  \n",
       "2                     0           0                     101             222  \n",
       "3                     0          47                      58             181  \n",
       "4                     0           0                      72             270  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#string_protein_id</th>\n",
       "      <th>alias</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>2B6H</td>\n",
       "      <td>BLAST_UniProt_DR_PDB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>2B6H</td>\n",
       "      <td>Ensembl_HGNC_UniProt_ID(supplied_by_UniProt)_D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>2B6H</td>\n",
       "      <td>Ensembl_PDB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>2B6H</td>\n",
       "      <td>Ensembl_UniProt_DR_PDB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>381</td>\n",
       "      <td>BLAST_KEGG_GENEID</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     #string_protein_id alias  \\\n",
       "0  9606.ENSP00000000233  2B6H   \n",
       "1  9606.ENSP00000000233  2B6H   \n",
       "2  9606.ENSP00000000233  2B6H   \n",
       "3  9606.ENSP00000000233  2B6H   \n",
       "4  9606.ENSP00000000233   381   \n",
       "\n",
       "                                              source  \n",
       "0                               BLAST_UniProt_DR_PDB  \n",
       "1  Ensembl_HGNC_UniProt_ID(supplied_by_UniProt)_D...  \n",
       "2                                        Ensembl_PDB  \n",
       "3                             Ensembl_UniProt_DR_PDB  \n",
       "4                                  BLAST_KEGG_GENEID  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "string_human_df = pd.read_csv(os.path.join(\n",
    "    data_dir, 'STRING/9606.protein.links.full.v11.5.txt'), sep=' ')\n",
    "display(string_human_df.head())\n",
    "string_aliases = pd.read_csv(os.path.join(\n",
    "    data_dir, 'STRING/9606.protein.aliases.v11.5.txt'), sep='\\t')\n",
    "display(string_aliases.head())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Integration\n",
    "\n",
    "To create a more relevant network that can yield more valuable insights, we use additional information to describe the genes (nodes) and their relationships (edges). \n",
    "<br>\n",
    "<br>\n",
    "\n",
    "### Network building\n",
    "#### Nodes\n",
    "Nodes will be all genes present in the STRING dataframe. Thus, we commence by standardizing their naming with the HAGR database (Entrez naming).\n",
    "STRING protein identifiers are encoded in the format 9606.gene, where 9606 is the code for homo sapiens. We translate the ids through a dictionary, but this is not exhaustive as not all entries in the alias dataframe have an entry for EntrezGene. Thus, we use a function that filters possibilities down to the most likely. This process should suffice for our purpose; although it could be more precise with more data, we make use of the resources available."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>neighborhood_transferred</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>homology</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>coexpression_transferred</th>\n",
       "      <th>experiments</th>\n",
       "      <th>experiments_transferred</th>\n",
       "      <th>database</th>\n",
       "      <th>database_transferred</th>\n",
       "      <th>textmining</th>\n",
       "      <th>textmining_transferred</th>\n",
       "      <th>combined_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARF5</td>\n",
       "      <td>PDE1C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>54</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>103</td>\n",
       "      <td>85</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ARF5</td>\n",
       "      <td>PAK2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>180</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARF5</td>\n",
       "      <td>RAB36</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "      <td>0</td>\n",
       "      <td>152</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>101</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ARF5</td>\n",
       "      <td>RAPGEF1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>161</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>47</td>\n",
       "      <td>58</td>\n",
       "      <td>181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARF5</td>\n",
       "      <td>SUMO3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>63</td>\n",
       "      <td>0</td>\n",
       "      <td>213</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11938493</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>872</td>\n",
       "      <td>213</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11938494</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>899</td>\n",
       "      <td>152</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11938495</th>\n",
       "      <td>NaN</td>\n",
       "      <td>KRTAP19-2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>182</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11938496</th>\n",
       "      <td>NaN</td>\n",
       "      <td>OR4D6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>843</td>\n",
       "      <td>155</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11938497</th>\n",
       "      <td>NaN</td>\n",
       "      <td>OR7D4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>898</td>\n",
       "      <td>168</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>167</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11938498 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         protein1   protein2  neighborhood  neighborhood_transferred  fusion  \\\n",
       "0            ARF5      PDE1C             0                         0       0   \n",
       "1            ARF5       PAK2             0                         0       0   \n",
       "2            ARF5      RAB36             0                         0       0   \n",
       "3            ARF5    RAPGEF1             0                         0       0   \n",
       "4            ARF5      SUMO3             0                         0       0   \n",
       "...           ...        ...           ...                       ...     ...   \n",
       "11938493      NaN        NaN             0                         0       0   \n",
       "11938494      NaN        NaN             0                         0       0   \n",
       "11938495      NaN  KRTAP19-2             0                         0       0   \n",
       "11938496      NaN      OR4D6             0                         0       0   \n",
       "11938497      NaN      OR7D4             0                         0       0   \n",
       "\n",
       "          cooccurence  homology  coexpression  coexpression_transferred  \\\n",
       "0                   0         0             0                        54   \n",
       "1                   0         0             0                         0   \n",
       "2                   0         0             0                        62   \n",
       "3                   0         0             0                         0   \n",
       "4                   0         0            60                        63   \n",
       "...               ...       ...           ...                       ...   \n",
       "11938493            0       872           213                         0   \n",
       "11938494            0       899           152                         0   \n",
       "11938495            0         0           182                         0   \n",
       "11938496            0       843           155                         0   \n",
       "11938497            0       898           168                         0   \n",
       "\n",
       "          experiments  experiments_transferred  database  \\\n",
       "0                   0                        0         0   \n",
       "1                   0                      180         0   \n",
       "2                   0                      152         0   \n",
       "3                   0                      161         0   \n",
       "4                   0                      213         0   \n",
       "...               ...                      ...       ...   \n",
       "11938493            0                        0         0   \n",
       "11938494            0                        0         0   \n",
       "11938495            0                        0         0   \n",
       "11938496            0                        0         0   \n",
       "11938497            0                        0         0   \n",
       "\n",
       "          database_transferred  textmining  textmining_transferred  \\\n",
       "0                            0         103                      85   \n",
       "1                            0           0                      61   \n",
       "2                            0           0                     101   \n",
       "3                            0          47                      58   \n",
       "4                            0           0                      72   \n",
       "...                        ...         ...                     ...   \n",
       "11938493                     0           0                       0   \n",
       "11938494                     0           0                       0   \n",
       "11938495                     0           0                       0   \n",
       "11938496                     0           0                       0   \n",
       "11938497                     0           0                       0   \n",
       "\n",
       "          combined_score  \n",
       "0                    155  \n",
       "1                    197  \n",
       "2                    222  \n",
       "3                    181  \n",
       "4                    270  \n",
       "...                  ...  \n",
       "11938493             213  \n",
       "11938494             151  \n",
       "11938495             181  \n",
       "11938496             154  \n",
       "11938497             167  \n",
       "\n",
       "[11938498 rows x 16 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We add a check to ensure that, if re-run, this cell wont undo previous progress\n",
    "\n",
    "previously_run = False if '9606' in string_human_df['protein1'].iloc[0] else True\n",
    "if not previously_run:  # Filter the DataFrame\n",
    "    biomart_table = pd.read_csv(os.path.join(\n",
    "        data_dir, 'misc/mart_export.txt'), sep='\\t')\n",
    "\n",
    "    # We crete the dictionary for translating\n",
    "    protein_gene_dict = {\n",
    "        transcript: gene for transcript, gene in\n",
    "        zip(biomart_table['Protein stable ID'].to_list(), biomart_table['Gene stable ID'].to_list())}\n",
    "    gene_symbol_dict = {\n",
    "        gene: symbol for gene, symbol in\n",
    "        zip(biomart_table['Gene stable ID'].to_list(), biomart_table['Gene name'].to_list())}\n",
    "\n",
    "    # First, we reformat STRING proteins to drop taxonomy code\n",
    "    string_human_df['protein1'] = string_human_df['protein1'].map(\n",
    "        lambda string: string.split('.')[1])\n",
    "    string_human_df['protein2'] = string_human_df['protein2'].map(\n",
    "        lambda string: string.split('.')[1])\n",
    "\n",
    "    # Next, we translate to gene code, then to symbol\n",
    "    string_human_df['protein1'] = string_human_df['protein1'].map(\n",
    "        protein_gene_dict)\n",
    "    string_human_df['protein2'] = string_human_df['protein2'].map(\n",
    "        protein_gene_dict)\n",
    "\n",
    "    string_human_df['protein1'] = string_human_df['protein1'].map(\n",
    "        gene_symbol_dict)\n",
    "    string_human_df['protein2'] = string_human_df['protein2'].map(\n",
    "        gene_symbol_dict)\n",
    "string_human_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "biomart_table = 0"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we use [Gene2Vec](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-018-5370-x) embeddings to represent gene co-expression. The embeddings have been found to capture gene localization (tissues) and function (pathways) by using co-expression data from 984 GEO database datasets.\n",
    "\n",
    "Gene2Vec's [GitHub repository](https://github.com/jingcheng-du/Gene2vec) provides embeddings of genes and we can use those to calculate gene relatedness by identifying the cosine similarity between both embeddings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We first load in the embedding data using Gensim's KeyedVectors and generate a function that will fetch each embedding\n",
    "from gensim.models import KeyedVectors\n",
    "\n",
    "\n",
    "def find_alternative_gene_name(gene, gene_symbol_dict_rev, protein_gene_dict_rev, aliases_by_protein):\n",
    "    \"\"\"\n",
    "    This function aims to replace gene names previously translated to those gene names found in the \n",
    "    gene2vec naming convention.\n",
    "\n",
    "    Args:\n",
    "        gene (str): gene synmbol\n",
    "        gene_symbol_dict_rev (dict): dictionary translating from gene symbol to Entrez gene ID\n",
    "        protein_gene_dict_rev (dict): dictionary translating from Entrez gene ID to protein\n",
    "        aliases_df (pd.DataFrame): dataframe on which to look for alternative names\n",
    "\n",
    "    Returns:\n",
    "        str: gene name present in g2v embedding object or original name\n",
    "    \"\"\"\n",
    "    if gene in gene_embeddings.key_to_index:  # Changed this line\n",
    "        return gene\n",
    "    if gene in gene_symbol_dict_rev:\n",
    "        gene = gene_symbol_dict_rev[gene]\n",
    "\n",
    "    protein = protein_gene_dict_rev.get(gene, None)\n",
    "    if protein:\n",
    "        alias_genes = aliases_by_protein.get(protein, [])\n",
    "        for alias in alias_genes:\n",
    "            if alias in gene_embeddings:\n",
    "                return alias\n",
    "\n",
    "    return gene\n",
    "\n",
    "\n",
    "def cosine_similarity(gene1, gene2):\n",
    "    \"\"\"\n",
    "    Takes in two gene embedding arrays and calculates the cosine similarity\n",
    "    between the vectors.\n",
    "    ! This function is equivalent to gensim's .similarity !\n",
    "    Args:\n",
    "        gene1 (arr): array of embeddings for the first gene\n",
    "        gene2 (arr): array of embeddings for the first gene\n",
    "\n",
    "    Returns:\n",
    "        float : cosine similarity between both vectors ranging from -1 to 1\n",
    "    \"\"\"\n",
    "    try:\n",
    "        vec1 = gene_embeddings[gene1]\n",
    "        vec2 = gene_embeddings[gene2]\n",
    "        return vec1.dot(vec2) / (np.linalg.norm(vec1) * np.linalg.norm(vec2))\n",
    "    except:\n",
    "        return np.nan\n",
    "\n",
    "\n",
    "def cosine_distance(gene1_in, gene2_in):\n",
    "    \"\"\"\n",
    "    Takes in two gene embedding arrays and calculates the cosine similarity and distance distance\n",
    "    between the vectors.\n",
    "    ! This function is equivalent to gensim's .distance !\n",
    "    Args:\n",
    "        gene1 (arr): array of embeddings for the first gene\n",
    "        gene2 (arr): array of embeddings for the first gene\n",
    "\n",
    "    Returns:\n",
    "        float : cosine similarity between both vectors ranging from -1 to 1\n",
    "        float : cosine similarity between both vectors ranging from 0 to 2\n",
    "    \"\"\"\n",
    "    gene1, gene2 = alternative_gene_name_dict[gene1_in], alternative_gene_name_dict[gene2_in]\n",
    "    try:\n",
    "\n",
    "        similarity = cosine_similarity(gene1, gene2)\n",
    "        if similarity == np.nan:\n",
    "            raise 'ERROR: GENE EMBEDDINGS NOT FOUND'\n",
    "        distance = 1 - similarity\n",
    "        return pd.Series([similarity, distance])\n",
    "    except:\n",
    "        return pd.Series([np.nan, np.nan])\n",
    "\n",
    "\n",
    "def process_batch(batch_df):\n",
    "    batch_df[['g2v_cossim', 'g2v_cosdist']] = batch_df[['protein1', 'protein2']].apply(\n",
    "        lambda row: cosine_distance(row['protein1'], row['protein2']), axis=1, result_type='expand')\n",
    "    return batch_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n",
      "3\n",
      "4\n",
      "  protein1 protein2  neighborhood  neighborhood_transferred  fusion  \\\n",
      "0     ARF5    PDE1C             0                         0       0   \n",
      "1     ARF5     PAK2             0                         0       0   \n",
      "2     ARF5    RAB36             0                         0       0   \n",
      "3     ARF5  RAPGEF1             0                         0       0   \n",
      "4     ARF5    SUMO3             0                         0       0   \n",
      "\n",
      "   cooccurence  homology  coexpression  coexpression_transferred  experiments  \\\n",
      "0            0         0             0                        54            0   \n",
      "1            0         0             0                         0            0   \n",
      "2            0         0             0                        62            0   \n",
      "3            0         0             0                         0            0   \n",
      "4            0         0            60                        63            0   \n",
      "\n",
      "   experiments_transferred  database  database_transferred  textmining  \\\n",
      "0                        0         0                     0         103   \n",
      "1                      180         0                     0           0   \n",
      "2                      152         0                     0           0   \n",
      "3                      161         0                     0          47   \n",
      "4                      213         0                     0           0   \n",
      "\n",
      "   textmining_transferred  combined_score  g2v_cossim  g2v_cosdist  \n",
      "0                      85             155   -0.066254     1.066254  \n",
      "1                      61             197    0.148998     0.851002  \n",
      "2                     101             222    0.098858     0.901142  \n",
      "3                      58             181    0.099199     0.900801  \n",
      "4                      72             270    0.198929     0.801071  \n"
     ]
    }
   ],
   "source": [
    "# Load the pre-trained gene embeddings\n",
    "gene_embeddings = KeyedVectors.load_word2vec_format(os.path.join(\n",
    "    data_dir, 'G2V/gene2vec_dim_200_iter_9_w2v.txt'), binary=False)\n",
    "\n",
    "# Translate the STRING gene names to those names which could be in the embedding data\n",
    "gene_symbol_dict_rev = {value: key for key, value in gene_symbol_dict.items()}\n",
    "string_aliases['sole_protein_id'] = string_aliases['#string_protein_id'].map(\n",
    "    lambda value: value.split('.')[1])\n",
    "protein_gene_dict_rev = {value: key for key,\n",
    "                         value in protein_gene_dict.items()}\n",
    "\n",
    "# Because iterating through all genes would be computationally redundant, we create a dictionary with corresponding gene names\n",
    "unique_genes = set(string_human_df['protein1']).union(\n",
    "    set(string_human_df['protein2']))\n",
    "aliases_by_protein = string_aliases.groupby(\n",
    "    'sole_protein_id')['alias'].apply(list).to_dict()\n",
    "\n",
    "alternative_gene_name_dict = {}\n",
    "for gene in unique_genes:\n",
    "    alternative_gene_name_dict[gene] = find_alternative_gene_name(\n",
    "        gene, gene_symbol_dict_rev, protein_gene_dict_rev, aliases_by_protein)\n",
    "\n",
    "# Obtain the embeddings for each pair and get the cosine similarity\n",
    "# Dropping NAs deletes about 900k pairs. We could receover them by integrating additional\n",
    "# database ID conversion information\n",
    "string_human_df = string_human_df.dropna()\n",
    "\n",
    "# !!!! CHANGE VARIABLE BELOW IF ALREADY RUN, CODE BELOW MAY BE TIME CONSUMING !!!!\n",
    "# You may also consider increasing the batch size if you're computational resources allow it\n",
    "previously_run = True\n",
    "if not previously_run:\n",
    "    # We process the data in batches due to memory restrictions\n",
    "\n",
    "    batch_size = 200000  # Adjust this value based on available memory\n",
    "    num_batches = int(np.ceil(len(string_human_df) / batch_size))\n",
    "\n",
    "    processed_batches = []\n",
    "\n",
    "    for i in tqdm(range(num_batches), desc=\"Processing batches\", unit=\"batch\"):\n",
    "        start = i * batch_size\n",
    "        end = (i + 1) * batch_size\n",
    "        batch_df = string_human_df[start:end].copy()\n",
    "        processed_batch = process_batch(batch_df)\n",
    "        processed_batches.append(processed_batch)\n",
    "\n",
    "    # Combine the processed batches\n",
    "    string_human_df_processed = pd.concat(processed_batches, ignore_index=True)\n",
    "    string_human_df_processed.to_csv(os.path.join(\n",
    "        data_dir, 'processed_string_hDf.csv'), index=False)\n",
    "\n",
    "else:\n",
    "    string_human_df_processed = pd.read_csv(\n",
    "        os.path.join(data_dir, 'processed_string_hDf.csv'))\n",
    "\n",
    "\n",
    "string_human_df_processed.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>neighborhood_transferred</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>homology</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>coexpression_transferred</th>\n",
       "      <th>experiments</th>\n",
       "      <th>experiments_transferred</th>\n",
       "      <th>database</th>\n",
       "      <th>database_transferred</th>\n",
       "      <th>textmining</th>\n",
       "      <th>textmining_transferred</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>g2v_cossim</th>\n",
       "      <th>g2v_cosdist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARF5</td>\n",
       "      <td>PDE1C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>54</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>103</td>\n",
       "      <td>85</td>\n",
       "      <td>155</td>\n",
       "      <td>-0.066254</td>\n",
       "      <td>1.066254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ARF5</td>\n",
       "      <td>PAK2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>180</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>197</td>\n",
       "      <td>0.148998</td>\n",
       "      <td>0.851002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARF5</td>\n",
       "      <td>RAB36</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "      <td>0</td>\n",
       "      <td>152</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>101</td>\n",
       "      <td>222</td>\n",
       "      <td>0.098858</td>\n",
       "      <td>0.901142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ARF5</td>\n",
       "      <td>RAPGEF1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>161</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>47</td>\n",
       "      <td>58</td>\n",
       "      <td>181</td>\n",
       "      <td>0.099199</td>\n",
       "      <td>0.900801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARF5</td>\n",
       "      <td>SUMO3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>63</td>\n",
       "      <td>0</td>\n",
       "      <td>213</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>270</td>\n",
       "      <td>0.198929</td>\n",
       "      <td>0.801071</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein1 protein2  neighborhood  neighborhood_transferred  fusion  \\\n",
       "0     ARF5    PDE1C             0                         0       0   \n",
       "1     ARF5     PAK2             0                         0       0   \n",
       "2     ARF5    RAB36             0                         0       0   \n",
       "3     ARF5  RAPGEF1             0                         0       0   \n",
       "4     ARF5    SUMO3             0                         0       0   \n",
       "\n",
       "   cooccurence  homology  coexpression  coexpression_transferred  experiments  \\\n",
       "0            0         0             0                        54            0   \n",
       "1            0         0             0                         0            0   \n",
       "2            0         0             0                        62            0   \n",
       "3            0         0             0                         0            0   \n",
       "4            0         0            60                        63            0   \n",
       "\n",
       "   experiments_transferred  database  database_transferred  textmining  \\\n",
       "0                        0         0                     0         103   \n",
       "1                      180         0                     0           0   \n",
       "2                      152         0                     0           0   \n",
       "3                      161         0                     0          47   \n",
       "4                      213         0                     0           0   \n",
       "\n",
       "   textmining_transferred  combined_score  g2v_cossim  g2v_cosdist  \n",
       "0                      85             155   -0.066254     1.066254  \n",
       "1                      61             197    0.148998     0.851002  \n",
       "2                     101             222    0.098858     0.901142  \n",
       "3                      58             181    0.099199     0.900801  \n",
       "4                      72             270    0.198929     0.801071  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "string_human_df_processed.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DEBUGGING\n",
    "The process above lead to some discrepancy in expected volume of data as demonstrated by the first code cell below, where the number of non-na values is much higher than it should (from ~2m to ~10m)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of common genes: 16353 of a total 18323 string genes\n",
      "Number of pairs with both genes in common genes: 9785708\n"
     ]
    }
   ],
   "source": [
    "# Get the unique genes in gene_embeddings\n",
    "genes_in_embeddings = set(gene_embeddings.vocab.keys())\n",
    "\n",
    "# Get the unique genes in string_human_df for protein1 and protein2\n",
    "unique_genes_string_human_df = set(string_human_df['protein1']).union(\n",
    "    set(string_human_df['protein2']))\n",
    "\n",
    "# Calculate the intersection of genes in gene_embeddings and string_human_df\n",
    "common_genes = genes_in_embeddings.intersection(unique_genes_string_human_df)\n",
    "\n",
    "# Print the number of common genes\n",
    "print(\n",
    "    f\"Number of common genes: {len(common_genes)} of a total {len(unique_genes_string_human_df)} string genes\")\n",
    "\n",
    "# Create a boolean mask to check if both genes in each pair are in common_genes\n",
    "both_genes_common_mask = string_human_df.apply(\n",
    "    lambda row: row['protein1'] in common_genes and row['protein2'] in common_genes, axis=1)\n",
    "\n",
    "# Count the number of pairs with both genes in common_genes\n",
    "both_genes_common_count = both_genes_common_mask.sum()\n",
    "\n",
    "print(\n",
    "    f\"Number of pairs with both genes in common genes: {both_genes_common_count}\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Above demonstrates that the number of non-na rows with data in both cosine distance and similarity should be at least 9m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/acc/miniconda3/envs/g2v/lib/python3.7/site-packages/ipykernel_launcher.py:9: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  if __name__ == \"__main__\":\n"
     ]
    },
    {
     "ename": "IndexingError",
     "evalue": "Unalignable boolean Series provided as indexer (index of the boolean Series and of the indexed object do not match).",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexingError\u001b[0m                             Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_420/3087921347.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;31m# Find the rows where both genes are in common genes, but the cosine similarity or distance is NaN\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0mdiscrepancy_mask\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mboth_genes_common_mask\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0mg2v_cosdist_nan_mask\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m \u001b[0mdiscrepancy_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstring_human_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdiscrepancy_mask\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"Number of discrepancies: {len(discrepancy_df)}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/g2v/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3447\u001b[0m         \u001b[0;31m# Do we have a (boolean) 1d indexer?\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3448\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_bool_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3449\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_bool_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3450\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3451\u001b[0m         \u001b[0;31m# We are left with two options: a single key, and a collection of keys,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/g2v/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m_getitem_bool_array\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3500\u001b[0m         \u001b[0;31m# check_bool_indexer will throw exception if Series key cannot\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3501\u001b[0m         \u001b[0;31m# be reindexed to match DataFrame rows\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3502\u001b[0;31m         \u001b[0mkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcheck_bool_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3503\u001b[0m         \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnonzero\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3504\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_take_with_is_copy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/g2v/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36mcheck_bool_indexer\u001b[0;34m(index, key)\u001b[0m\n\u001b[1;32m   2387\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mmask\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0many\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2388\u001b[0m             raise IndexingError(\n\u001b[0;32m-> 2389\u001b[0;31m                 \u001b[0;34m\"Unalignable boolean Series provided as \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2390\u001b[0m                 \u001b[0;34m\"indexer (index of the boolean Series and of \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2391\u001b[0m                 \u001b[0;34m\"the indexed object do not match).\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIndexingError\u001b[0m: Unalignable boolean Series provided as indexer (index of the boolean Series and of the indexed object do not match)."
     ]
    }
   ],
   "source": [
    "# Create a boolean mask for rows where both genes are in common genes\n",
    "both_genes_common_mask = string_human_df_processed.apply(\n",
    "    lambda row: row['protein1'] in common_genes and row['protein2'] in common_genes, axis=1)\n",
    "\n",
    "# Create a boolean mask for rows where the cosine similarity or distance is NaN\n",
    "g2v_cosdist_nan_mask = string_human_df_processed['g2v_cosdist'].isna()\n",
    "\n",
    "# Find the rows where both genes are in common genes, but the cosine similarity or distance is NaN\n",
    "discrepancy_mask = both_genes_common_mask & g2v_cosdist_nan_mask\n",
    "discrepancy_df = string_human_df[discrepancy_mask]\n",
    "\n",
    "print(f\"Number of discrepancies: {len(discrepancy_df)}\")\n",
    "print(\"Examples of discrepancies:\")\n",
    "print(discrepancy_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0662540671033174"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_embeddings.distance(string_human_df[['protein1', 'protein2']].iloc[0][0], string_human_df[[\n",
    "                         'protein1', 'protein2']].iloc[0][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2ML1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A3GALT2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19835</th>\n",
       "      <td>ZYG11A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19836</th>\n",
       "      <td>ZYG11B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19837</th>\n",
       "      <td>ZYX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19838</th>\n",
       "      <td>ZZEF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19839</th>\n",
       "      <td>ZZZ3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19840 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene\n",
       "0         A1BG\n",
       "1         A1CF\n",
       "2          A2M\n",
       "3        A2ML1\n",
       "4      A3GALT2\n",
       "...        ...\n",
       "19835   ZYG11A\n",
       "19836   ZYG11B\n",
       "19837      ZYX\n",
       "19838    ZZEF1\n",
       "19839     ZZZ3\n",
       "\n",
       "[19840 rows x 1 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Generate an object that contains node information\n",
    "node_characteristics_df = pd.DataFrame({\"gene\": sorted(\n",
    "    list(set([i for i in gene_symbol_dict.values() if type(i) == str])))})\n",
    "\n",
    "node_characteristics_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENSG00000198888': 'MT-ND1',\n",
       " 'ENSG00000198763': 'MT-ND2',\n",
       " 'ENSG00000198804': 'MT-CO1',\n",
       " 'ENSG00000198712': 'MT-CO2',\n",
       " 'ENSG00000228253': 'MT-ATP8',\n",
       " 'ENSG00000198899': 'MT-ATP6',\n",
       " 'ENSG00000198938': 'MT-CO3',\n",
       " 'ENSG00000198840': 'MT-ND3',\n",
       " 'ENSG00000212907': 'MT-ND4L',\n",
       " 'ENSG00000198886': 'MT-ND4',\n",
       " 'ENSG00000198786': 'MT-ND5',\n",
       " 'ENSG00000198695': 'MT-ND6',\n",
       " 'ENSG00000198727': 'MT-CYB',\n",
       " 'ENSG00000278704': nan,\n",
       " 'ENSG00000271254': nan,\n",
       " 'ENSG00000281486': 'SNTG2',\n",
       " 'ENSG00000273735': 'KIR3DL2',\n",
       " 'ENSG00000280663': 'PCMTD2',\n",
       " 'ENSG00000262826': 'INTS3',\n",
       " 'ENSG00000276379': 'KIR3DL1',\n",
       " 'ENSG00000276433': 'KIR3DL3',\n",
       " 'ENSG00000273947': 'KIR2DL3',\n",
       " 'ENSG00000276518': 'KIR3DP1',\n",
       " 'ENSG00000276595': 'PTCHD3',\n",
       " 'ENSG00000278152': 'KIR2DS2',\n",
       " 'ENSG00000277044': 'OPRL1',\n",
       " 'ENSG00000277339': 'NPBWR2',\n",
       " 'ENSG00000276044': 'KIR2DL4',\n",
       " 'ENSG00000277554': 'KIR2DL3',\n",
       " 'ENSG00000276779': 'KIR2DL4',\n",
       " 'ENSG00000275974': 'PPIAL4F',\n",
       " 'ENSG00000276885': 'KIR2DS4',\n",
       " 'ENSG00000274714': 'KIR2DS4',\n",
       " 'ENSG00000276534': 'KIR3DS1',\n",
       " 'ENSG00000277650': 'KIR2DS5',\n",
       " 'ENSG00000277334': 'NCR1',\n",
       " 'ENSG00000277484': 'KIR2DL3',\n",
       " 'ENSG00000273661': 'KIR2DL2',\n",
       " 'ENSG00000275047': 'KIR2DS5',\n",
       " 'ENSG00000276333': 'HS3ST6',\n",
       " 'ENSG00000275237': 'KIR2DL4',\n",
       " 'ENSG00000277078': 'KIR2DS4',\n",
       " 'ENSG00000274807': 'KIR2DS4',\n",
       " 'ENSG00000261992': 'GATAD2B',\n",
       " 'ENSG00000276858': 'FCAR',\n",
       " 'ENSG00000275659': 'KIR3DL1',\n",
       " 'ENSG00000275407': 'KIR2DL2',\n",
       " 'ENSG00000274696': 'KIR3DL3',\n",
       " 'ENSG00000278805': 'KIR2DL1',\n",
       " 'ENSG00000274324': 'KIR2DS4',\n",
       " 'ENSG00000277345': 'KIR2DS4',\n",
       " 'ENSG00000274676': 'KIR2DL5A',\n",
       " 'ENSG00000275699': 'KIR2DL4',\n",
       " 'ENSG00000273510': 'KIR2DL1',\n",
       " 'ENSG00000276820': 'KIR2DL1',\n",
       " 'ENSG00000273575': 'KIR2DL4',\n",
       " 'ENSG00000278788': 'SBNO2',\n",
       " 'ENSG00000278074': 'KIR2DL4',\n",
       " 'ENSG00000278306': 'KIR2DS3',\n",
       " 'ENSG00000273526': 'KIR2DS4',\n",
       " 'ENSG00000278731': 'KIR2DL2',\n",
       " 'ENSG00000275433': 'KIR3DL3',\n",
       " 'ENSG00000276876': 'MYT1',\n",
       " 'ENSG00000280830': 'ADI1',\n",
       " 'ENSG00000262647': 'OR9G4',\n",
       " 'ENSG00000263163': 'SLC27A3',\n",
       " 'ENSG00000281107': 'OR5G3',\n",
       " 'ENSG00000261958': 'OR9G1',\n",
       " 'ENSG00000274782': 'KIR2DL1',\n",
       " 'ENSG00000275018': nan,\n",
       " 'ENSG00000262191': 'OR9G9',\n",
       " 'ENSG00000274891': 'TRAPPC12',\n",
       " 'ENSG00000276424': 'KIR3DL2',\n",
       " 'ENSG00000277317': 'KIR2DL3',\n",
       " 'ENSG00000273775': 'KIR3DL1',\n",
       " 'ENSG00000269823': nan,\n",
       " 'ENSG00000275080': 'KIR2DL1',\n",
       " 'ENSG00000276676': 'KIR2DS5',\n",
       " 'ENSG00000275511': 'KIR3DL2',\n",
       " 'ENSG00000277620': 'KIR3DL3',\n",
       " 'ENSG00000276328': 'KIR3DL3',\n",
       " 'ENSG00000276258': 'KIR2DS2',\n",
       " 'ENSG00000275946': 'KIR2DL5B',\n",
       " 'ENSG00000276423': 'KIR3DL1',\n",
       " 'ENSG00000274406': 'KIR2DS4',\n",
       " 'ENSG00000274189': 'KIR2DL4',\n",
       " 'ENSG00000277725': 'KIR2DL2',\n",
       " 'ENSG00000277616': 'KIR2DL1',\n",
       " 'ENSG00000277355': 'KIR2DL4',\n",
       " 'ENSG00000278369': 'KIR2DL3',\n",
       " 'ENSG00000276930': 'KIR3DL3',\n",
       " 'ENSG00000275521': 'NCR1',\n",
       " 'ENSG00000277272': 'KIR3DL1',\n",
       " 'ENSG00000276425': 'KIR2DS2',\n",
       " 'ENSG00000276086': 'KIR3DL3',\n",
       " 'ENSG00000275421': 'KIR2DS1',\n",
       " 'ENSG00000274948': 'KIR3DL1',\n",
       " 'ENSG00000273603': 'KIR2DS1',\n",
       " 'ENSG00000279566': 'ZNF43',\n",
       " 'ENSG00000275596': 'KIR2DL5A',\n",
       " 'ENSG00000273738': 'FCAR',\n",
       " 'ENSG00000274830': 'KIR2DL3',\n",
       " 'ENSG00000275276': 'KIR2DL1',\n",
       " 'ENSG00000274707': 'KIR2DL5B',\n",
       " 'ENSG00000274746': 'ZNF100',\n",
       " 'ENSG00000274193': 'KIR2DL4',\n",
       " 'ENSG00000275732': 'KIR2DL4',\n",
       " 'ENSG00000281763': 'DHX36',\n",
       " 'ENSG00000273517': 'KIR2DS1',\n",
       " 'ENSG00000275786': 'KIR3DL1',\n",
       " 'ENSG00000278780': 'ZNF430',\n",
       " 'ENSG00000275172': 'KIR3DL3',\n",
       " 'ENSG00000277193': 'KIR2DS3',\n",
       " 'ENSG00000275288': 'KIR3DL1',\n",
       " 'ENSG00000276392': 'KIR2DL5A',\n",
       " 'ENSG00000275717': 'KIR3DL1',\n",
       " 'ENSG00000278850': 'KIR3DL2',\n",
       " 'ENSG00000273578': 'KIR2DL2',\n",
       " 'ENSG00000274722': 'KIR3DL2',\n",
       " 'ENSG00000273887': 'KIR2DL3',\n",
       " 'ENSG00000273502': 'KIR3DL3',\n",
       " 'ENSG00000274955': 'KIR2DL4',\n",
       " 'ENSG00000277101': 'ARHGEF26',\n",
       " 'ENSG00000276450': 'NCR1',\n",
       " 'ENSG00000276501': 'KIR3DL1',\n",
       " 'ENSG00000278091': 'ZNF85',\n",
       " 'ENSG00000274518': 'KIR2DS2',\n",
       " 'ENSG00000276329': 'KIR3DL1',\n",
       " 'ENSG00000276218': 'KIR2DL3',\n",
       " 'ENSG00000277964': 'KIR2DL4',\n",
       " 'ENSG00000276590': 'KIR2DL3',\n",
       " 'ENSG00000267728': 'KRTAP9-9',\n",
       " 'ENSG00000274945': 'KIR2DL4',\n",
       " 'ENSG00000263090': 'KRTAP9-2',\n",
       " 'ENSG00000262512': 'KRTAP9-4',\n",
       " 'ENSG00000278710': 'KIR3DL2',\n",
       " 'ENSG00000278755': 'KIR2DL1',\n",
       " 'ENSG00000262749': 'KRTAP9-3',\n",
       " 'ENSG00000267635': nan,\n",
       " 'ENSG00000262218': 'KRTAP29-1',\n",
       " 'ENSG00000274480': 'KIR3DL3',\n",
       " 'ENSG00000275452': 'KIR2DS2',\n",
       " 'ENSG00000278474': 'KIR3DL2',\n",
       " 'ENSG00000274108': 'KIR2DL3',\n",
       " 'ENSG00000275050': 'ZNF66',\n",
       " 'ENSG00000276004': 'KIR3DL2',\n",
       " 'ENSG00000276806': 'KIR3DL3',\n",
       " 'ENSG00000277596': 'KIR3DL3',\n",
       " 'ENSG00000274724': 'KIR3DL3',\n",
       " 'ENSG00000274610': 'KIR3DP1',\n",
       " 'ENSG00000278327': 'KIR2DL3',\n",
       " 'ENSG00000274243': 'KIR2DS5',\n",
       " 'ENSG00000274563': 'KIR2DS5',\n",
       " 'ENSG00000276985': 'FCAR',\n",
       " 'ENSG00000278414': 'KIR2DL5B',\n",
       " 'ENSG00000281020': 'SNTG2',\n",
       " 'ENSG00000275608': 'KIR3DS1',\n",
       " 'ENSG00000275750': 'KIR2DL1',\n",
       " 'ENSG00000274429': 'DLG5',\n",
       " 'ENSG00000275970': 'FCAR',\n",
       " 'ENSG00000277362': 'KIR2DL4',\n",
       " 'ENSG00000277356': 'KIR2DL1',\n",
       " 'ENSG00000276395': 'KIR2DS4',\n",
       " 'ENSG00000276498': 'KIR3DS1',\n",
       " 'ENSG00000278726': 'KIR3DL2',\n",
       " 'ENSG00000274556': 'KIR3DL3',\n",
       " 'ENSG00000277196': 'PRODH',\n",
       " 'ENSG00000277603': 'TPO',\n",
       " 'ENSG00000273518': 'KIR3DL1',\n",
       " 'ENSG00000278817': nan,\n",
       " 'ENSG00000276017': nan,\n",
       " 'ENSG00000277630': nan,\n",
       " 'ENSG00000275269': 'FCAR',\n",
       " 'ENSG00000276310': 'KIR2DL1',\n",
       " 'ENSG00000275136': 'FCAR',\n",
       " 'ENSG00000278442': 'KIR3DL2',\n",
       " 'ENSG00000275051': 'PIWIL1',\n",
       " 'ENSG00000277750': 'KIR2DL4',\n",
       " 'ENSG00000262573': 'KRTAP9-2',\n",
       " 'ENSG00000262896': 'KRTAP9-4',\n",
       " 'ENSG00000274382': 'PROP1',\n",
       " 'ENSG00000263263': 'KRTAP9-6',\n",
       " 'ENSG00000276465': 'KIR2DS4',\n",
       " 'ENSG00000262221': 'KRTAP9-7',\n",
       " 'ENSG00000275914': 'KIR2DL2',\n",
       " 'ENSG00000262542': 'KRTAP29-1',\n",
       " 'ENSG00000282665': 'ERICH1',\n",
       " 'ENSG00000262594': 'KRTAP9-9',\n",
       " 'ENSG00000275937': 'KIR2DL5B',\n",
       " 'ENSG00000277885': 'KIR2DS2',\n",
       " 'ENSG00000276011': 'KIR2DL2',\n",
       " 'ENSG00000281935': 'DNAH10',\n",
       " 'ENSG00000273794': 'KIR2DL1',\n",
       " 'ENSG00000273498': 'KIR2DL4',\n",
       " 'ENSG00000278277': nan,\n",
       " 'ENSG00000278229': 'RPS17',\n",
       " 'ENSG00000278758': 'KIR3DL2',\n",
       " 'ENSG00000278692': 'KIR2DL2',\n",
       " 'ENSG00000274786': 'KIR3DL3',\n",
       " 'ENSG00000275735': 'KIR2DS2',\n",
       " 'ENSG00000280635': 'PROP1',\n",
       " 'ENSG00000278079': 'KIR3DL1',\n",
       " 'ENSG00000274947': 'KIR2DS4',\n",
       " 'ENSG00000278207': 'KIR2DL1',\n",
       " 'ENSG00000276256': nan,\n",
       " 'ENSG00000277354': 'CRIM1',\n",
       " 'ENSG00000277392': 'KIR3DL3',\n",
       " 'ENSG00000281775': 'FTCD',\n",
       " 'ENSG00000275262': 'KIR3DL2',\n",
       " 'ENSG00000277552': 'KIR3DL3',\n",
       " 'ENSG00000275035': 'CCDC92',\n",
       " 'ENSG00000278025': 'NCR1',\n",
       " 'ENSG00000276459': 'KIR2DL3',\n",
       " 'ENSG00000279389': 'GARRE1',\n",
       " 'ENSG00000277850': 'KIR2DL4',\n",
       " 'ENSG00000282752': 'CTDP1',\n",
       " 'ENSG00000262102': 'DSEL',\n",
       " 'ENSG00000275517': 'PRB1',\n",
       " 'ENSG00000274957': 'KIR2DS4',\n",
       " 'ENSG00000275629': 'KIR3DL2',\n",
       " 'ENSG00000262634': 'SKA1',\n",
       " 'ENSG00000275353': 'KIR2DS4',\n",
       " 'ENSG00000282302': 'SEPTIN9',\n",
       " 'ENSG00000275416': 'KIR3DL2',\n",
       " 'ENSG00000275848': 'KIR2DL4',\n",
       " 'ENSG00000277442': 'NCR1',\n",
       " 'ENSG00000280807': 'TUBGCP5',\n",
       " 'ENSG00000262860': 'LSM14A',\n",
       " 'ENSG00000281289': 'LSS',\n",
       " 'ENSG00000278248': 'KIR2DL1',\n",
       " 'ENSG00000275637': 'NCR1',\n",
       " 'ENSG00000277028': 'KIR3DL3',\n",
       " 'ENSG00000274679': 'SPATC1L',\n",
       " 'ENSG00000280618': 'CYFIP1',\n",
       " 'ENSG00000275436': 'KIR2DL5B',\n",
       " 'ENSG00000278247': 'GOLGA6L1',\n",
       " 'ENSG00000274465': 'KIR3DS1',\n",
       " 'ENSG00000262156': 'APOBEC3A',\n",
       " 'ENSG00000275009': 'GOLGA6L1',\n",
       " 'ENSG00000281919': 'SLC45A2',\n",
       " 'ENSG00000277069': 'RXFP3',\n",
       " 'ENSG00000281690': 'ADAMTS12',\n",
       " 'ENSG00000278728': 'CNTNAP2',\n",
       " 'ENSG00000282152': 'DLGAP2',\n",
       " 'ENSG00000278809': 'KIR3DL2',\n",
       " 'ENSG00000276357': 'KIR3DL2',\n",
       " 'ENSG00000277445': 'CPEB1',\n",
       " 'ENSG00000280759': 'AP2A2',\n",
       " 'ENSG00000278495': 'KIR2DL1',\n",
       " 'ENSG00000276582': 'ABCB11',\n",
       " 'ENSG00000275306': 'KIR2DS1',\n",
       " 'ENSG00000278368': 'KIR3DL1',\n",
       " 'ENSG00000278373': 'G6PC2',\n",
       " 'ENSG00000261846': 'AADACL2',\n",
       " 'ENSG00000274639': 'KIR3DL3',\n",
       " 'ENSG00000273772': 'SMN2',\n",
       " 'ENSG00000276882': 'KIR3DL2',\n",
       " 'ENSG00000276154': 'KIR2DS4',\n",
       " 'ENSG00000278116': 'KIR2DL5A',\n",
       " 'ENSG00000276625': 'KIR2DL1',\n",
       " 'ENSG00000278430': 'KIR2DL4',\n",
       " 'ENSG00000275500': nan,\n",
       " 'ENSG00000275031': 'METRNL',\n",
       " 'ENSG00000276327': 'KIR2DS1',\n",
       " 'ENSG00000277909': 'SPC25',\n",
       " 'ENSG00000274320': 'GOLGA6L9',\n",
       " 'ENSG00000262473': 'GART',\n",
       " 'ENSG00000282755': nan,\n",
       " 'ENSG00000274042': 'GOLGA6L10',\n",
       " 'ENSG00000274920': 'KIR3DL1',\n",
       " 'ENSG00000275049': nan,\n",
       " 'ENSG00000262911': 'DNAJC28',\n",
       " 'ENSG00000277518': 'MUC6',\n",
       " 'ENSG00000262795': 'IFNGR2',\n",
       " 'ENSG00000277657': 'KIR2DS3',\n",
       " 'ENSG00000278403': 'KIR3DL2',\n",
       " 'ENSG00000274166': 'GOLGA6L10',\n",
       " 'ENSG00000275737': 'KIR2DS2',\n",
       " 'ENSG00000278481': 'KIR2DL5B',\n",
       " 'ENSG00000274296': 'DECR2',\n",
       " 'ENSG00000276856': 'TUBGCP5',\n",
       " 'ENSG00000277163': 'KIR2DS3',\n",
       " 'ENSG00000278759': 'TBCD',\n",
       " 'ENSG00000275326': 'NOSTRIN',\n",
       " 'ENSG00000278086': nan,\n",
       " 'ENSG00000277540': 'KIR2DL4',\n",
       " 'ENSG00000278821': 'KIR2DL1',\n",
       " 'ENSG00000275513': 'KIR3DL3',\n",
       " 'ENSG00000274438': 'KIR2DS2',\n",
       " 'ENSG00000275317': 'KIR2DL4',\n",
       " 'ENSG00000273831': 'LRRC56',\n",
       " 'ENSG00000274692': 'KIR2DL1',\n",
       " 'ENSG00000274404': 'GOLGA6L22',\n",
       " 'ENSG00000263057': 'KRT20',\n",
       " 'ENSG00000262068': 'KRTAP3-1',\n",
       " 'ENSG00000277389': 'KRTAP5-6',\n",
       " 'ENSG00000280585': 'KRTAP5-1',\n",
       " 'ENSG00000278620': 'KRTAP5-4',\n",
       " 'ENSG00000263296': 'KRTAP3-2',\n",
       " 'ENSG00000262121': 'KRTAP4-9',\n",
       " 'ENSG00000263091': 'KRTAP2-4',\n",
       " 'ENSG00000278302': 'KRTAP5-5',\n",
       " 'ENSG00000278165': 'DUSP8',\n",
       " 'ENSG00000281216': 'KIF26B',\n",
       " 'ENSG00000275062': 'KIR3DL3',\n",
       " 'ENSG00000281618': 'IFITM10',\n",
       " 'ENSG00000281191': 'KRTAP5-3',\n",
       " 'ENSG00000281533': 'KRTAP5-2',\n",
       " 'ENSG00000278856': 'KIR3DL1',\n",
       " 'ENSG00000263309': 'KRT23',\n",
       " 'ENSG00000275338': 'RAB11FIP3',\n",
       " 'ENSG00000263160': 'TMEM50B',\n",
       " 'ENSG00000278361': 'KIR3DL2',\n",
       " 'ENSG00000277033': 'B3GNTL1',\n",
       " 'ENSG00000262845': 'KRT40',\n",
       " 'ENSG00000268674': nan,\n",
       " 'ENSG00000277475': nan,\n",
       " 'ENSG00000278633': nan,\n",
       " 'ENSG00000263141': 'PAXBP1',\n",
       " 'ENSG00000262808': 'KRTAP1-1',\n",
       " 'ENSG00000262731': nan,\n",
       " 'ENSG00000263229': 'KRTAP4-11',\n",
       " 'ENSG00000262938': 'C21orf62',\n",
       " 'ENSG00000275456': 'KIR2DL4',\n",
       " 'ENSG00000262347': 'KRTAP1-3',\n",
       " 'ENSG00000277982': 'KIR3DL2',\n",
       " 'ENSG00000276910': 'GTF2H2',\n",
       " 'ENSG00000262520': 'KRTAP4-16',\n",
       " 'ENSG00000281758': 'KCNIP4',\n",
       " 'ENSG00000263243': 'KRT12',\n",
       " 'ENSG00000275008': 'KIR2DL3',\n",
       " 'ENSG00000275037': 'KIR3DS1',\n",
       " 'ENSG00000261862': 'KRTAP1-5',\n",
       " 'ENSG00000275599': 'KIR2DS3',\n",
       " 'ENSG00000262164': 'KRT39',\n",
       " 'ENSG00000261852': 'KRTAP2-1',\n",
       " 'ENSG00000276731': 'KIR2DL2',\n",
       " 'ENSG00000263101': 'KRTAP3-3',\n",
       " 'ENSG00000262862': 'KRTAP2-3',\n",
       " 'ENSG00000263236': 'KRTAP4-12',\n",
       " 'ENSG00000276177': 'KIR2DL5A',\n",
       " 'ENSG00000263242': 'KRTAP1-4',\n",
       " 'ENSG00000266286': nan,\n",
       " 'ENSG00000261968': 'KRTAP4-8',\n",
       " 'ENSG00000274780': 'PHRF1',\n",
       " 'ENSG00000277848': 'MEGF11',\n",
       " 'ENSG00000274394': 'KIR3DL3',\n",
       " 'ENSG00000277175': 'KIR3DL1',\n",
       " 'ENSG00000274283': 'KIR3DS1',\n",
       " 'ENSG00000275522': 'KIR2DL1',\n",
       " 'ENSG00000278304': 'KIR2DS1',\n",
       " 'ENSG00000277429': 'SERF1B',\n",
       " 'ENSG00000276686': 'KIR2DL5A',\n",
       " 'ENSG00000274763': 'KIR3DL3',\n",
       " 'ENSG00000275623': 'KIR2DL3',\n",
       " 'ENSG00000277251': 'KIR2DL2',\n",
       " 'ENSG00000276896': 'GOLGA8F',\n",
       " 'ENSG00000275873': 'LMNTD2',\n",
       " 'ENSG00000277365': 'NDUFA6',\n",
       " 'ENSG00000274112': 'SMDT1',\n",
       " 'ENSG00000275430': 'MUC20',\n",
       " 'ENSG00000272835': 'SMDT1',\n",
       " 'ENSG00000273397': 'NDUFA6',\n",
       " 'ENSG00000276068': 'NAIP',\n",
       " 'ENSG00000278201': 'KIR2DL4',\n",
       " 'ENSG00000275211': 'CYP2D6',\n",
       " 'ENSG00000272000': nan,\n",
       " 'ENSG00000275428': nan,\n",
       " 'ENSG00000272765': 'NDUFA6',\n",
       " 'ENSG00000273651': 'GOLGA8G',\n",
       " 'ENSG00000273936': 'TNFRSF14',\n",
       " 'ENSG00000276016': 'ABR',\n",
       " 'ENSG00000276230': 'RNH1',\n",
       " 'ENSG00000276371': 'SINHCAF',\n",
       " 'ENSG00000262024': 'TCF20',\n",
       " 'ENSG00000281742': 'MCCC2',\n",
       " 'ENSG00000272532': 'CYP2D6',\n",
       " 'ENSG00000276461': 'TCF20',\n",
       " 'ENSG00000272901': 'SMDT1',\n",
       " 'ENSG00000281151': 'KCNQ2',\n",
       " 'ENSG00000278427': 'KIR3DL1',\n",
       " 'ENSG00000275253': 'KIR2DS2',\n",
       " 'ENSG00000277278': 'HERC2',\n",
       " 'ENSG00000274700': nan,\n",
       " 'ENSG00000278606': 'KIR2DL4',\n",
       " 'ENSG00000278088': 'CYP2D7',\n",
       " 'ENSG00000275571': 'TTC34',\n",
       " 'ENSG00000263181': 'CYP2D7',\n",
       " 'ENSG00000277131': 'MMEL1',\n",
       " 'ENSG00000276825': 'DRD4',\n",
       " 'ENSG00000223898': 'OR11A1',\n",
       " 'ENSG00000275626': 'KIR3DL2',\n",
       " 'ENSG00000226832': 'OR2B3',\n",
       " 'ENSG00000274803': 'BDP1',\n",
       " 'ENSG00000282712': 'DEAF1',\n",
       " 'ENSG00000278501': 'TIMM22',\n",
       " 'ENSG00000281300': 'NXN',\n",
       " 'ENSG00000278114': 'MUC20',\n",
       " 'ENSG00000276084': 'KIR3DL3',\n",
       " 'ENSG00000275351': 'KIR2DS4',\n",
       " 'ENSG00000278659': 'OR2T6',\n",
       " 'ENSG00000275754': 'OR2T2',\n",
       " 'ENSG00000233636': 'OR2J3',\n",
       " 'ENSG00000275617': 'OR2T4',\n",
       " 'ENSG00000243441': 'OR5V1',\n",
       " 'ENSG00000281489': 'OR2T7',\n",
       " 'ENSG00000276561': 'IRF7',\n",
       " 'ENSG00000273508': 'OR2T1',\n",
       " 'ENSG00000277113': 'OR2T3',\n",
       " 'ENSG00000234100': 'OR14J1',\n",
       " 'ENSG00000275102': 'OR2T5',\n",
       " 'ENSG00000237258': 'OR11A1',\n",
       " 'ENSG00000278271': 'KIR2DL4',\n",
       " 'ENSG00000276798': 'LILRA4',\n",
       " 'ENSG00000275349': 'SMN1',\n",
       " 'ENSG00000239741': 'LY6G6F',\n",
       " 'ENSG00000233095': 'HLA-G',\n",
       " 'ENSG00000281038': 'OR8U9',\n",
       " 'ENSG00000263150': 'OR8J2',\n",
       " 'ENSG00000231094': 'PSORS1C1',\n",
       " 'ENSG00000277753': 'MUC20',\n",
       " 'ENSG00000277833': 'KIR2DL1',\n",
       " 'ENSG00000262315': 'OR8U8',\n",
       " 'ENSG00000261897': 'OR5T3',\n",
       " 'ENSG00000228652': 'OR2W1',\n",
       " 'ENSG00000262851': 'OR5T2',\n",
       " 'ENSG00000262755': 'OR8K3',\n",
       " 'ENSG00000274254': 'KIR3DL3',\n",
       " 'ENSG00000274402': 'KIR2DL3',\n",
       " 'ENSG00000262611': 'OR8H1',\n",
       " 'ENSG00000274971': nan,\n",
       " 'ENSG00000278489': 'KRTAP5-2',\n",
       " 'ENSG00000262784': 'OR5T1',\n",
       " 'ENSG00000275731': 'KIR2DS4',\n",
       " 'ENSG00000263328': 'OR8K1',\n",
       " 'ENSG00000275920': 'KRTAP5-3',\n",
       " 'ENSG00000277068': 'KRTAP5-4',\n",
       " 'ENSG00000262796': 'OR8J1',\n",
       " 'ENSG00000275581': 'SERF1A',\n",
       " 'ENSG00000278210': 'KRTAP5-1',\n",
       " 'ENSG00000236560': 'PPP1R11',\n",
       " 'ENSG00000277291': 'KRTAP5-5',\n",
       " 'ENSG00000184494': 'NEU1',\n",
       " 'ENSG00000273748': nan,\n",
       " 'ENSG00000273793': 'DUSP8',\n",
       " 'ENSG00000223687': 'ZNF311',\n",
       " 'ENSG00000206470': 'MAS1L',\n",
       " 'ENSG00000276565': nan,\n",
       " 'ENSG00000206467': 'OR2H2',\n",
       " 'ENSG00000276163': 'LAIR1',\n",
       " 'ENSG00000237634': 'LINC01556',\n",
       " 'ENSG00000261893': 'CLK2',\n",
       " 'ENSG00000235487': 'PSORS1C1',\n",
       " 'ENSG00000223448': 'LTB',\n",
       " 'ENSG00000223852': 'ZFP57',\n",
       " 'ENSG00000276848': 'KIR3DX1',\n",
       " 'ENSG00000232178': 'OR2J3',\n",
       " 'ENSG00000235754': 'APOM',\n",
       " 'ENSG00000273494': 'PANK4',\n",
       " 'ENSG00000231319': 'OR2B3',\n",
       " 'ENSG00000231041': 'LINC01556',\n",
       " 'ENSG00000276536': 'HRAS',\n",
       " 'ENSG00000260873': 'SNTB2',\n",
       " 'ENSG00000275486': 'KIR3DL1',\n",
       " 'ENSG00000274646': 'RPS9',\n",
       " 'ENSG00000239836': 'PSMB9',\n",
       " 'ENSG00000137394': 'TRIM10',\n",
       " 'ENSG00000273859': 'RASSF7',\n",
       " 'ENSG00000276139': 'KIR2DS2',\n",
       " 'ENSG00000262785': 'PKLR',\n",
       " 'ENSG00000262788': 'UTP4',\n",
       " 'ENSG00000274513': 'LILRB2',\n",
       " 'ENSG00000278384': nan,\n",
       " 'ENSG00000275118': 'MBOAT7',\n",
       " 'ENSG00000276429': 'PLCH2',\n",
       " 'ENSG00000229470': 'GNL1',\n",
       " 'ENSG00000230995': 'PPP1R10',\n",
       " 'ENSG00000226858': 'ZFP57',\n",
       " 'ENSG00000274084': 'LAIR2',\n",
       " 'ENSG00000273529': 'HES5',\n",
       " 'ENSG00000263290': 'SCAMP3',\n",
       " 'ENSG00000276066': 'VSTM1',\n",
       " 'ENSG00000233813': 'MRPS18B',\n",
       " 'ENSG00000274638': 'NLRP2',\n",
       " 'ENSG00000276613': 'MUC4',\n",
       " 'ENSG00000263238': 'CTSO',\n",
       " 'ENSG00000278216': 'UGT2A3',\n",
       " 'ENSG00000262505': 'ASIC5',\n",
       " 'ENSG00000275083': 'KIR3DL2',\n",
       " 'ENSG00000262635': 'TDO2',\n",
       " 'ENSG00000278555': 'LILRB4',\n",
       " 'ENSG00000275125': 'PRXL2B',\n",
       " 'ENSG00000275190': 'UGT2B10',\n",
       " 'ENSG00000274941': 'CNOT3',\n",
       " 'ENSG00000233312': 'C4B_2',\n",
       " 'ENSG00000236843': 'NRM',\n",
       " 'ENSG00000231325': 'LY6G5C',\n",
       " 'ENSG00000276061': 'NDUFA3',\n",
       " 'ENSG00000277132': 'UGT2B15',\n",
       " 'ENSG00000274146': 'KIR3DL1',\n",
       " 'ENSG00000231257': 'ATAT1',\n",
       " 'ENSG00000277400': nan,\n",
       " 'ENSG00000274847': 'MAFIP',\n",
       " 'ENSG00000237724': 'HSPA1A',\n",
       " 'ENSG00000276628': 'LENG1',\n",
       " 'ENSG00000273572': 'CDHR5',\n",
       " 'ENSG00000276766': 'NDUFA3',\n",
       " 'ENSG00000278300': 'KIR2DS2',\n",
       " 'ENSG00000236251': 'HSPA1L',\n",
       " 'ENSG00000262446': 'GBA1',\n",
       " 'ENSG00000274058': 'TMPRSS11E',\n",
       " 'ENSG00000275333': 'FAM120B',\n",
       " 'ENSG00000233490': 'GPSM3',\n",
       " 'ENSG00000276452': 'LILRB1',\n",
       " 'ENSG00000273506': nan,\n",
       " 'ENSG00000274473': 'SCT',\n",
       " 'ENSG00000276979': 'KIR2DL4',\n",
       " 'ENSG00000278490': 'KIR3DL3',\n",
       " 'ENSG00000234101': 'OR2W1',\n",
       " 'ENSG00000282367': 'PTDSS2',\n",
       " 'ENSG00000275555': 'DLL1',\n",
       " 'ENSG00000274905': 'TMC4',\n",
       " 'ENSG00000236250': 'STK19',\n",
       " 'ENSG00000276341': 'RDH13',\n",
       " 'ENSG00000274511': 'KIR3DL3',\n",
       " 'ENSG00000226463': 'OR2W1',\n",
       " 'ENSG00000236927': 'OR14J1',\n",
       " 'ENSG00000275584': 'LILRA6',\n",
       " 'ENSG00000274078': 'TSEN34',\n",
       " 'ENSG00000236826': 'LSM2',\n",
       " 'ENSG00000263324': 'HCN3',\n",
       " 'ENSG00000274185': 'LILRA4',\n",
       " 'ENSG00000278355': 'LILRA5',\n",
       " 'ENSG00000248496': 'ZBED9',\n",
       " 'ENSG00000235966': 'OR12D2',\n",
       " 'ENSG00000206450': 'HLA-B',\n",
       " 'ENSG00000233046': 'OR5V1',\n",
       " 'ENSG00000282185': 'PTPRN2',\n",
       " 'ENSG00000277359': 'RPS9',\n",
       " 'ENSG00000229071': 'PRRT1',\n",
       " 'ENSG00000278519': 'MBOAT7',\n",
       " 'ENSG00000224708': 'LY6G6E',\n",
       " 'ENSG00000234495': 'TRIM27',\n",
       " 'ENSG00000275272': 'YTHDC1',\n",
       " 'ENSG00000262599': 'DENND11',\n",
       " 'ENSG00000229861': 'MRPS18B',\n",
       " 'ENSG00000239927': 'RPP21',\n",
       " 'ENSG00000273822': 'MUC4',\n",
       " 'ENSG00000204692': 'OR12D3',\n",
       " 'ENSG00000279552': 'ZNF729',\n",
       " 'ENSG00000196231': 'OR2J2',\n",
       " 'ENSG00000233687': 'OR2B3',\n",
       " 'ENSG00000225351': 'ZNF311',\n",
       " 'ENSG00000276219': 'ZNF676',\n",
       " 'ENSG00000277179': 'NLRP7',\n",
       " 'ENSG00000274000': 'LILRA2',\n",
       " 'ENSG00000233054': 'OR2B3',\n",
       " 'ENSG00000262666': 'ENTREP3',\n",
       " 'ENSG00000227420': 'MRPS18B',\n",
       " 'ENSG00000278270': 'RPS9',\n",
       " 'ENSG00000278468': 'MUC4',\n",
       " 'ENSG00000276920': 'NKD2',\n",
       " 'ENSG00000277076': 'KIR2DL4',\n",
       " 'ENSG00000231488': 'ABHD16A',\n",
       " 'ENSG00000238289': 'MICB',\n",
       " 'ENSG00000278533': 'OSCAR',\n",
       " 'ENSG00000278362': 'NCR1',\n",
       " 'ENSG00000233439': 'PSORS1C1',\n",
       " 'ENSG00000241863': 'RPP21',\n",
       " 'ENSG00000224472': 'TCF19',\n",
       " 'ENSG00000234343': 'NEU1',\n",
       " 'ENSG00000112459': 'OR14J1',\n",
       " 'ENSG00000277006': 'DEFB104B',\n",
       " 'ENSG00000275196': 'KIR2DL1',\n",
       " 'ENSG00000276482': 'SLC12A7',\n",
       " 'ENSG00000225845': 'BTNL2',\n",
       " 'ENSG00000278670': 'GP6',\n",
       " 'ENSG00000274729': 'DEFB105B',\n",
       " 'ENSG00000273884': 'LILRA3',\n",
       " 'ENSG00000262771': 'SSBP1',\n",
       " 'ENSG00000274540': 'PRR23D1',\n",
       " 'ENSG00000276263': 'TFPT',\n",
       " 'ENSG00000275444': 'DEFB4B',\n",
       " 'ENSG00000262217': 'SLC5A8',\n",
       " 'ENSG00000234951': 'SAPCD1',\n",
       " 'ENSG00000276504': 'TFPT',\n",
       " 'ENSG00000273641': 'DEFB103B',\n",
       " 'ENSG00000275546': 'KIR2DL2',\n",
       " 'ENSG00000278154': 'LAIR1',\n",
       " 'ENSG00000278613': 'NAIP',\n",
       " 'ENSG00000276209': 'KIR2DS4',\n",
       " 'ENSG00000262139': 'ANO4',\n",
       " 'ENSG00000274174': 'NLRP7',\n",
       " 'ENSG00000276146': 'LILRB2',\n",
       " 'ENSG00000274195': 'ZNF705G',\n",
       " 'ENSG00000224544': 'PSORS1C2',\n",
       " 'ENSG00000277530': 'DEFB107B',\n",
       " 'ENSG00000274651': 'PRPF31',\n",
       " 'ENSG00000262327': 'AGK',\n",
       " 'ENSG00000277178': 'TARM1',\n",
       " 'ENSG00000237403': 'PPP1R11',\n",
       " 'ENSG00000230791': 'LST1',\n",
       " 'ENSG00000275918': 'KIR3DX1',\n",
       " 'ENSG00000235915': 'LST1',\n",
       " 'ENSG00000274405': 'SPAG11B',\n",
       " 'ENSG00000274131': 'CEACAM4',\n",
       " 'ENSG00000278303': 'MUC4',\n",
       " 'ENSG00000224994': 'TRIM39',\n",
       " 'ENSG00000204690': 'OR12D2',\n",
       " 'ENSG00000276463': 'TAF9',\n",
       " 'ENSG00000263042': 'WEE2',\n",
       " 'ENSG00000278565': 'CEACAM21',\n",
       " 'ENSG00000277114': 'CNOT3',\n",
       " 'ENSG00000275794': 'DEFB106B',\n",
       " 'ENSG00000231021': 'HLA-DRB4',\n",
       " 'ENSG00000275658': 'KIR2DL3',\n",
       " 'ENSG00000273953': 'OR10A3',\n",
       " 'ENSG00000276025': 'OR5P2',\n",
       " 'ENSG00000278253': 'OR5P3',\n",
       " 'ENSG00000276451': 'OR10A6',\n",
       " 'ENSG00000226271': 'OR2J3',\n",
       " 'ENSG00000276387': 'KIR2DS1',\n",
       " 'ENSG00000276780': 'NLRP10',\n",
       " 'ENSG00000278656': 'KIR3DL2',\n",
       " 'ENSG00000280501': 'CEACAM7',\n",
       " 'ENSG00000237051': 'GABBR1',\n",
       " 'ENSG00000278634': 'LILRA2',\n",
       " 'ENSG00000277263': 'HLA-DRA',\n",
       " 'ENSG00000275798': 'BTNL2',\n",
       " 'ENSG00000231543': 'C2',\n",
       " 'ENSG00000275434': 'KIR3DS1',\n",
       " 'ENSG00000274739': 'KIR2DS5',\n",
       " 'ENSG00000231179': 'MICB',\n",
       " 'ENSG00000275583': nan,\n",
       " 'ENSG00000243958': 'PSMB9',\n",
       " 'ENSG00000276464': 'POMZP3',\n",
       " 'ENSG00000275463': 'LILRB2',\n",
       " 'ENSG00000273984': 'MUC4',\n",
       " 'ENSG00000236808': 'POLR1H',\n",
       " 'ENSG00000282147': 'FAM20C',\n",
       " 'ENSG00000278415': 'FCAR',\n",
       " 'ENSG00000276184': 'UPK3B',\n",
       " 'ENSG00000226437': 'TRIM39',\n",
       " 'ENSG00000277177': 'LILRA6',\n",
       " 'ENSG00000235769': 'LY6G6E',\n",
       " 'ENSG00000237703': 'TRIM10',\n",
       " 'ENSG00000262145': 'ATP23',\n",
       " 'ENSG00000281950': 'SPDYE16',\n",
       " 'ENSG00000275838': 'KIR3DL2',\n",
       " 'ENSG00000278120': 'KIR2DS1',\n",
       " 'ENSG00000236967': 'RNF39',\n",
       " 'ENSG00000282379': 'DTX2',\n",
       " 'ENSG00000223757': 'VWA7',\n",
       " 'ENSG00000231350': 'MUC21',\n",
       " 'ENSG00000274412': 'KIR2DL2',\n",
       " 'ENSG00000228978': 'TNF',\n",
       " 'ENSG00000232289': 'OR11A1',\n",
       " 'ENSG00000206471': 'OR2H1',\n",
       " 'ENSG00000226634': 'DDAH2',\n",
       " 'ENSG00000276420': 'PRAMEF5',\n",
       " 'ENSG00000277853': 'PRAMEF18',\n",
       " 'ENSG00000276345': nan,\n",
       " 'ENSG00000276572': 'KIR3DL3',\n",
       " 'ENSG00000278021': 'PRAMEF8',\n",
       " 'ENSG00000274921': 'KIR2DS4',\n",
       " 'ENSG00000277800': 'TBX22',\n",
       " 'ENSG00000274109': 'HNRNPCL2',\n",
       " 'ENSG00000273911': 'KIR3DL2',\n",
       " 'ENSG00000276299': 'PRAMEF25',\n",
       " 'ENSG00000273865': 'HNRNPCL4',\n",
       " 'ENSG00000276370': 'PRAMEF33',\n",
       " 'ENSG00000274843': 'MUC20',\n",
       " 'ENSG00000281573': 'PRAMEF9',\n",
       " 'ENSG00000273943': 'CNOT3',\n",
       " 'ENSG00000278707': 'KIR3DL2',\n",
       " 'ENSG00000282711': 'TRBV11-1',\n",
       " 'ENSG00000206455': 'TCF19',\n",
       " 'ENSG00000274889': 'TARM1',\n",
       " 'ENSG00000282704': 'TRBV7-6',\n",
       " 'ENSG00000278426': 'LENG1',\n",
       " 'ENSG00000282618': 'TRBV10-1',\n",
       " 'ENSG00000282756': 'TRBV7-4',\n",
       " 'ENSG00000282577': 'TRBV5-5',\n",
       " 'ENSG00000259101': 'LY6G6F-LY6G6D',\n",
       " 'ENSG00000276342': 'TMEM179',\n",
       " 'ENSG00000282242': 'TRBV11-3',\n",
       " 'ENSG00000235155': 'B3GALT4',\n",
       " 'ENSG00000282621': 'TRBV19',\n",
       " 'ENSG00000282354': 'TRBV12-4',\n",
       " 'ENSG00000282225': 'TRBV7-1',\n",
       " 'ENSG00000281963': 'TRBV25-1',\n",
       " 'ENSG00000282483': 'TRBV17',\n",
       " 'ENSG00000282148': 'TRBV5-3',\n",
       " 'ENSG00000281943': 'TRBV7-9',\n",
       " 'ENSG00000282748': 'TRBV5-7',\n",
       " 'ENSG00000282234': 'TRBV27',\n",
       " 'ENSG00000282506': 'TRBV7-2',\n",
       " 'ENSG00000277816': 'LILRB3',\n",
       " 'ENSG00000274571': 'NLRP7',\n",
       " 'ENSG00000274143': 'KIR2DL5A',\n",
       " 'ENSG00000172850': 'LSM2',\n",
       " 'ENSG00000277071': 'NLRP7',\n",
       " 'ENSG00000231321': 'RXRB',\n",
       " 'ENSG00000274126': 'C14orf180',\n",
       " 'ENSG00000223680': 'DDR1',\n",
       " 'ENSG00000278046': 'LILRA3',\n",
       " 'ENSG00000274538': 'PHLDA2',\n",
       " 'ENSG00000278466': 'MUC2',\n",
       " 'ENSG00000230855': 'OR2J3',\n",
       " 'ENSG00000232367': 'TAP1',\n",
       " 'ENSG00000282291': 'SLC39A11',\n",
       " 'ENSG00000234549': 'ATAT1',\n",
       " 'ENSG00000276159': 'VSTM1',\n",
       " 'ENSG00000281385': 'AP2A2',\n",
       " 'ENSG00000274161': 'DLGAP2',\n",
       " 'ENSG00000274348': 'TRPV5',\n",
       " 'ENSG00000275404': 'LILRA5',\n",
       " 'ENSG00000227715': 'HLA-A',\n",
       " 'ENSG00000275399': 'NLRP2',\n",
       " 'ENSG00000282605': 'TRBV12-5',\n",
       " 'ENSG00000282589': 'TRBJ2-2P',\n",
       " 'ENSG00000281957': 'TRBJ1-6',\n",
       " 'ENSG00000282215': 'TRBJ1-5',\n",
       " 'ENSG00000275577': 'VSTM1',\n",
       " 'ENSG00000282064': 'TRBV20-1',\n",
       " 'ENSG00000206372': 'C2',\n",
       " 'ENSG00000282543': nan,\n",
       " 'ENSG00000229300': 'SNHG32',\n",
       " 'ENSG00000282007': 'TRBV10-2',\n",
       " 'ENSG00000274953': 'VSTM1',\n",
       " 'ENSG00000262607': 'IQSEC3',\n",
       " 'ENSG00000275821': 'LLCFC1',\n",
       " 'ENSG00000242386': 'HLA-DMB',\n",
       " 'ENSG00000282204': 'TRBV9',\n",
       " 'ENSG00000242038': 'EGFL8',\n",
       " 'ENSG00000281981': 'TRBC1',\n",
       " 'ENSG00000281630': 'MUC20',\n",
       " 'ENSG00000282488': 'TRBJ2-1',\n",
       " 'ENSG00000274566': 'GP6',\n",
       " 'ENSG00000272803': 'PRH2',\n",
       " 'ENSG00000282676': nan,\n",
       " 'ENSG00000275644': 'OSCAR',\n",
       " 'ENSG00000262111': 'TAS2R30',\n",
       " 'ENSG00000263028': 'TAS2R19',\n",
       " 'ENSG00000230332': 'RNF39',\n",
       " 'ENSG00000272712': 'TAS2R8',\n",
       " 'ENSG00000274050': 'GP6',\n",
       " 'ENSG00000273326': 'TAS2R7',\n",
       " 'ENSG00000273431': 'TAS2R50',\n",
       " 'ENSG00000277414': 'LILRB5',\n",
       " 'ENSG00000262612': 'TAS2R43',\n",
       " 'ENSG00000234096': 'MOG',\n",
       " 'ENSG00000261984': 'TAS2R14',\n",
       " 'ENSG00000277281': 'MUC6',\n",
       " 'ENSG00000273086': 'TAS2R9',\n",
       " 'ENSG00000263097': 'TAS2R31',\n",
       " 'ENSG00000277807': 'LILRB1',\n",
       " 'ENSG00000273457': 'TAS2R13',\n",
       " 'ENSG00000262525': 'TAS2R46',\n",
       " 'ENSG00000277923': 'MBOAT7',\n",
       " 'ENSG00000234218': 'MICB',\n",
       " 'ENSG00000273092': 'TAS2R20',\n",
       " 'ENSG00000274796': 'TSEN34',\n",
       " 'ENSG00000277824': nan,\n",
       " 'ENSG00000276971': 'TRPV6',\n",
       " 'ENSG00000275724': 'NDUFA3',\n",
       " 'ENSG00000274332': 'LRRC37A2',\n",
       " 'ENSG00000230293': 'MSH5',\n",
       " 'ENSG00000235692': 'PFDN6',\n",
       " 'ENSG00000226651': 'CLIC1',\n",
       " 'ENSG00000263247': 'PRR4',\n",
       " 'ENSG00000282181': 'MROH5',\n",
       " 'ENSG00000278081': 'RPS9',\n",
       " 'ENSG00000275633': 'GP6',\n",
       " 'ENSG00000227357': 'HLA-DRB4',\n",
       " 'ENSG00000282470': 'TRBV6-7',\n",
       " 'ENSG00000275474': 'RDH13',\n",
       " 'ENSG00000233618': 'OR2J3',\n",
       " 'ENSG00000275551': 'OSCAR',\n",
       " 'ENSG00000282013': 'RPH3AL',\n",
       " 'ENSG00000282049': 'PRSS2',\n",
       " 'ENSG00000282203': 'TRBV7-3',\n",
       " 'ENSG00000282771': 'TRBV18',\n",
       " 'ENSG00000281484': 'BTG3',\n",
       " 'ENSG00000276323': 'TFPT',\n",
       " 'ENSG00000275575': 'PTP4A3',\n",
       " 'ENSG00000278608': 'MAMDC2',\n",
       " 'ENSG00000275885': 'PRPF31',\n",
       " 'ENSG00000276220': 'NDUFA3',\n",
       " 'ENSG00000239463': 'HLA-DMA',\n",
       " 'ENSG00000277751': 'LILRB2',\n",
       " 'ENSG00000274311': 'LILRB5',\n",
       " 'ENSG00000272805': 'TAS2R10',\n",
       " 'ENSG00000276826': 'RDH13',\n",
       " 'ENSG00000274247': 'PRSS1',\n",
       " 'ENSG00000230279': 'LTA',\n",
       " 'ENSG00000275181': 'GPR20',\n",
       " 'ENSG00000206245': 'TSBP1',\n",
       " 'ENSG00000274914': 'LILRA5',\n",
       " 'ENSG00000278789': 'NLRP2',\n",
       " 'ENSG00000235986': 'GNL1',\n",
       " 'ENSG00000277776': 'NLRP7',\n",
       " 'ENSG00000278322': 'MBOAT7',\n",
       " 'ENSG00000282449': 'TRBV23-1',\n",
       " 'ENSG00000227858': 'ZFP57',\n",
       " 'ENSG00000276033': 'KIR3DX1',\n",
       " 'ENSG00000282040': nan,\n",
       " 'ENSG00000282466': 'TRBV5-4',\n",
       " 'ENSG00000225550': 'OR2J2',\n",
       " 'ENSG00000276618': 'RAD17',\n",
       " 'ENSG00000277733': 'MBOAT7',\n",
       " 'ENSG00000273564': 'ENTREP2',\n",
       " 'ENSG00000277378': 'OR6V1',\n",
       " 'ENSG00000275796': 'NLRP2',\n",
       " 'ENSG00000282273': 'TRBJ1-4',\n",
       " 'ENSG00000277667': 'TMC4',\n",
       " 'ENSG00000264619': 'MSH5-SAPCD1',\n",
       " 'ENSG00000275165': 'TSEN34',\n",
       " 'ENSG00000206461': 'C6orf15',\n",
       " 'ENSG00000278624': 'AZU1',\n",
       " 'ENSG00000276296': 'TFPT',\n",
       " 'ENSG00000281078': 'IGKV1D-8',\n",
       " 'ENSG00000278563': 'MGAM2',\n",
       " 'ENSG00000274005': 'RPS9',\n",
       " 'ENSG00000276495': 'APBA2',\n",
       " 'ENSG00000275777': 'IGKV3D-7',\n",
       " 'ENSG00000281759': nan,\n",
       " 'ENSG00000224587': 'MDC1',\n",
       " 'ENSG00000275117': 'PRPF31',\n",
       " 'ENSG00000277439': 'GP6',\n",
       " 'ENSG00000273833': 'TFPT',\n",
       " 'ENSG00000273579': 'C22orf15',\n",
       " 'ENSG00000206409': 'APOM',\n",
       " 'ENSG00000273914': 'OR9A2',\n",
       " 'ENSG00000274669': 'LILRB1',\n",
       " 'ENSG00000282407': 'TRBV13',\n",
       " 'ENSG00000275539': 'LILRA6',\n",
       " 'ENSG00000282812': 'TRBV28',\n",
       " 'ENSG00000282272': 'TRBJ2-2',\n",
       " 'ENSG00000230780': 'OR11A1',\n",
       " 'ENSG00000275979': 'CNOT3',\n",
       " 'ENSG00000277804': 'PRTN3',\n",
       " 'ENSG00000276051': 'HLA-G',\n",
       " 'ENSG00000282019': 'GPI',\n",
       " 'ENSG00000282283': 'PLPPR3',\n",
       " 'ENSG00000224264': 'VARS1',\n",
       " 'ENSG00000277953': 'PRPF31',\n",
       " 'ENSG00000274176': 'CNOT3',\n",
       " 'ENSG00000274273': 'ATP9B',\n",
       " 'ENSG00000206383': 'HSPA1L',\n",
       " 'ENSG00000275960': 'KIR2DL2',\n",
       " 'ENSG00000233795': 'POLR1H',\n",
       " 'ENSG00000263310': 'SALL3',\n",
       " 'ENSG00000282813': 'ARMC10',\n",
       " 'ENSG00000277273': 'CDK7',\n",
       " 'ENSG00000273896': 'TSEN34',\n",
       " 'ENSG00000276774': 'LENG1',\n",
       " 'ENSG00000275384': 'TARM1',\n",
       " 'ENSG00000278723': 'KIR3DL3',\n",
       " 'ENSG00000276421': 'PRPF31',\n",
       " 'ENSG00000275123': 'TARM1',\n",
       " 'ENSG00000277730': 'SCGB1C2',\n",
       " 'ENSG00000227222': 'DHX16',\n",
       " 'ENSG00000275837': 'SMARCB1',\n",
       " 'ENSG00000274626': 'RPS9',\n",
       " 'ENSG00000282092': 'MED16',\n",
       " 'ENSG00000275482': 'EPHB6',\n",
       " 'ENSG00000232804': 'HSPA1B',\n",
       " 'ENSG00000276260': 'TMC4',\n",
       " 'ENSG00000273916': nan,\n",
       " 'ENSG00000274703': 'OSCAR',\n",
       " 'ENSG00000215641': 'TRIM27',\n",
       " 'ENSG00000276458': 'LENG8',\n",
       " 'ENSG00000274504': 'RDH13',\n",
       " 'ENSG00000263014': 'ATP6V1G3',\n",
       " 'ENSG00000278191': 'CARS1',\n",
       " 'ENSG00000276802': 'HERC2',\n",
       " 'ENSG00000277571': 'ELANE',\n",
       " 'ENSG00000275723': 'NAPEPLD',\n",
       " 'ENSG00000277457': 'LENG1',\n",
       " 'ENSG00000273992': 'NLRP2',\n",
       " 'ENSG00000228520': 'RING1',\n",
       " 'ENSG00000273607': 'CHCHD10',\n",
       " 'ENSG00000276155': 'MAPT',\n",
       " 'ENSG00000277996': 'TMC4',\n",
       " 'ENSG00000239974': 'EGFL8',\n",
       " 'ENSG00000278279': 'LILRB4',\n",
       " 'ENSG00000278365': 'NDUFA3',\n",
       " 'ENSG00000282017': nan,\n",
       " 'ENSG00000274533': 'KIR2DS4',\n",
       " 'ENSG00000262334': 'RPH3AL',\n",
       " 'ENSG00000277110': 'TRBV6-5',\n",
       " 'ENSG00000282252': 'TRBV14',\n",
       " 'ENSG00000274619': 'CFD',\n",
       " 'ENSG00000282285': 'TRBV4-2',\n",
       " 'ENSG00000235441': 'OR10C1',\n",
       " 'ENSG00000274926': 'KIR2DL1',\n",
       " 'ENSG00000282054': nan,\n",
       " 'ENSG00000282803': 'TRBV5-1',\n",
       " 'ENSG00000282353': nan,\n",
       " 'ENSG00000282154': 'TRBJ2-6',\n",
       " 'ENSG00000282179': 'TRBV7-7',\n",
       " 'ENSG00000206256': 'PPT2',\n",
       " 'ENSG00000282735': 'GARRE1',\n",
       " 'ENSG00000282568': 'TRBV2',\n",
       " 'ENSG00000277722': 'NDUFA3',\n",
       " 'ENSG00000206430': 'NCR3',\n",
       " 'ENSG00000277090': 'GOLGA8G',\n",
       " 'ENSG00000226165': 'HLA-DQB2',\n",
       " 'ENSG00000275365': 'MMP11',\n",
       " 'ENSG00000235067': 'TUBB',\n",
       " 'ENSG00000277615': 'CNOT3',\n",
       " 'ENSG00000282307': 'SERPINF1',\n",
       " 'ENSG00000267954': nan,\n",
       " 'ENSG00000228859': 'LY6G6C',\n",
       " 'ENSG00000274437': 'DERL3',\n",
       " 'ENSG00000255419': 'MSH5-SAPCD1',\n",
       " 'ENSG00000278257': 'LENG1',\n",
       " 'ENSG00000276022': 'TFPT',\n",
       " 'ENSG00000282607': 'MGAM',\n",
       " 'ENSG00000272670': 'DOC2B',\n",
       " 'ENSG00000282415': 'TRBV16',\n",
       " 'ENSG00000282076': 'KCNQ1',\n",
       " 'ENSG00000281970': 'TRBV6-1',\n",
       " 'ENSG00000277111': 'PLEKHM1',\n",
       " 'ENSG00000234704': 'BRD2',\n",
       " 'ENSG00000282604': 'CCL4L2',\n",
       " 'ENSG00000262418': 'PTPRC',\n",
       " 'ENSG00000274144': 'PRPF31',\n",
       " 'ENSG00000276785': 'PRSS58',\n",
       " 'ENSG00000282730': 'TRBV24-1',\n",
       " 'ENSG00000282392': 'TRBJ2-7',\n",
       " 'ENSG00000282500': 'TRBJ1-3',\n",
       " 'ENSG00000282610': nan,\n",
       " 'ENSG00000278284': 'RDH13',\n",
       " 'ENSG00000282014': 'TRBV4-1',\n",
       " 'ENSG00000276849': 'TRBC2',\n",
       " 'ENSG00000277361': 'OCA2',\n",
       " 'ENSG00000282340': 'TRBV10-3',\n",
       " 'ENSG00000282078': 'TRBJ1-1',\n",
       " 'ENSG00000279643': 'TAMM41',\n",
       " 'ENSG00000282371': 'TRBJ2-3',\n",
       " 'ENSG00000282820': 'TRBJ2-5',\n",
       " 'ENSG00000243009': 'RPP21',\n",
       " 'ENSG00000282098': 'TRBV5-6',\n",
       " 'ENSG00000282240': 'TRBV6-4',\n",
       " 'ENSG00000281989': 'TRBJ1-2',\n",
       " 'ENSG00000276537': 'TTYH1',\n",
       " 'ENSG00000282253': 'TRBD1',\n",
       " 'ENSG00000282380': 'TRBJ2-4',\n",
       " 'ENSG00000282719': 'TRBV6-2',\n",
       " 'ENSG00000206495': 'TRIM39',\n",
       " 'ENSG00000282297': 'TRBV30',\n",
       " 'ENSG00000282208': 'TRBV12-3',\n",
       " 'ENSG00000282130': 'NSMCE3',\n",
       " 'ENSG00000282624': 'TRBV3-1',\n",
       " 'ENSG00000282134': 'TRBV6-8',\n",
       " 'ENSG00000282628': 'TRBV29-1',\n",
       " 'ENSG00000282459': 'TRBV6-6',\n",
       " 'ENSG00000282497': 'TRBV15',\n",
       " 'ENSG00000277015': 'SALL3',\n",
       " 'ENSG00000275483': 'NLRP7',\n",
       " 'ENSG00000282035': nan,\n",
       " 'ENSG00000274575': 'IGLV1-40',\n",
       " 'ENSG00000229006': 'TRIM27',\n",
       " 'ENSG00000276739': 'KIR3DL2',\n",
       " 'ENSG00000274854': 'IGLV1-44',\n",
       " 'ENSG00000276336': 'SCARF1',\n",
       " 'ENSG00000280866': 'IGLV3-32',\n",
       " 'ENSG00000277421': 'ZNF280B',\n",
       " 'ENSG00000282266': 'ATP9B',\n",
       " 'ENSG00000204487': 'LTB',\n",
       " 'ENSG00000275917': 'CHRFAM7A',\n",
       " 'ENSG00000278062': 'ZNF251',\n",
       " 'ENSG00000281471': 'IGLV5-37',\n",
       " 'ENSG00000278293': 'IGLV7-43',\n",
       " 'ENSG00000273737': 'IGLV1-36',\n",
       " 'ENSG00000274053': nan,\n",
       " 'ENSG00000275499': 'ZNF280A',\n",
       " 'ENSG00000281536': nan,\n",
       " 'ENSG00000281345': 'IGLV2-33',\n",
       " 'ENSG00000281145': 'IGLV5-45',\n",
       " 'ENSG00000274734': 'ARHGAP11B',\n",
       " 'ENSG00000276196': 'KIR3DL3',\n",
       " 'ENSG00000276887': 'TTYH1',\n",
       " 'ENSG00000280688': 'CEP72',\n",
       " 'ENSG00000278682': 'NLRP2',\n",
       " 'ENSG00000281861': 'SLC9A3',\n",
       " 'ENSG00000275246': 'TLCD2',\n",
       " 'ENSG00000275013': 'PRAME',\n",
       " 'ENSG00000278363': 'TMC4',\n",
       " 'ENSG00000276615': 'KEL',\n",
       " 'ENSG00000231002': 'TRIM26',\n",
       " 'ENSG00000275861': 'TBC1D3I',\n",
       " 'ENSG00000276021': 'WDR81',\n",
       " 'ENSG00000278712': 'TSEN34',\n",
       " 'ENSG00000275501': 'MUC20',\n",
       " 'ENSG00000276468': 'KIR2DP1',\n",
       " 'ENSG00000277709': 'KIR3DL2',\n",
       " 'ENSG00000275568': 'ARHGAP11A',\n",
       " ...}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_symbol_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Protein stable ID</th>\n",
       "      <th>Reactome ID</th>\n",
       "      <th>NCBI gene (formerly Entrezgene) ID</th>\n",
       "      <th>BioGRID Interaction data, The General Repository for Interaction Datasets ID</th>\n",
       "      <th>Gene name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>ENSP00000354687</td>\n",
       "      <td>R-HSA-1430728</td>\n",
       "      <td>4535.0</td>\n",
       "      <td>110631.0</td>\n",
       "      <td>MT-ND1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>ENSP00000354687</td>\n",
       "      <td>R-HSA-1428517</td>\n",
       "      <td>4535.0</td>\n",
       "      <td>110631.0</td>\n",
       "      <td>MT-ND1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>ENSP00000354687</td>\n",
       "      <td>R-HSA-163200</td>\n",
       "      <td>4535.0</td>\n",
       "      <td>110631.0</td>\n",
       "      <td>MT-ND1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>ENSP00000354687</td>\n",
       "      <td>R-HSA-611105</td>\n",
       "      <td>4535.0</td>\n",
       "      <td>110631.0</td>\n",
       "      <td>MT-ND1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>ENSP00000354687</td>\n",
       "      <td>R-HSA-6799198</td>\n",
       "      <td>4535.0</td>\n",
       "      <td>110631.0</td>\n",
       "      <td>MT-ND1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>519070</th>\n",
       "      <td>ENSG00000162437</td>\n",
       "      <td>ENSP00000397069</td>\n",
       "      <td>NaN</td>\n",
       "      <td>55225.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RAVER2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>519071</th>\n",
       "      <td>ENSG00000122432</td>\n",
       "      <td>ENSP00000514413</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100505741.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPATA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>519072</th>\n",
       "      <td>ENSG00000122432</td>\n",
       "      <td>ENSP00000514414</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100505741.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPATA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>519073</th>\n",
       "      <td>ENSG00000122432</td>\n",
       "      <td>ENSP00000514416</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100505741.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPATA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>519074</th>\n",
       "      <td>ENSG00000122432</td>\n",
       "      <td>ENSP00000514417</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100505741.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPATA1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>519075 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene stable ID Protein stable ID    Reactome ID  \\\n",
       "0       ENSG00000198888   ENSP00000354687  R-HSA-1430728   \n",
       "1       ENSG00000198888   ENSP00000354687  R-HSA-1428517   \n",
       "2       ENSG00000198888   ENSP00000354687   R-HSA-163200   \n",
       "3       ENSG00000198888   ENSP00000354687   R-HSA-611105   \n",
       "4       ENSG00000198888   ENSP00000354687  R-HSA-6799198   \n",
       "...                 ...               ...            ...   \n",
       "519070  ENSG00000162437   ENSP00000397069            NaN   \n",
       "519071  ENSG00000122432   ENSP00000514413            NaN   \n",
       "519072  ENSG00000122432   ENSP00000514414            NaN   \n",
       "519073  ENSG00000122432   ENSP00000514416            NaN   \n",
       "519074  ENSG00000122432   ENSP00000514417            NaN   \n",
       "\n",
       "        NCBI gene (formerly Entrezgene) ID  \\\n",
       "0                                   4535.0   \n",
       "1                                   4535.0   \n",
       "2                                   4535.0   \n",
       "3                                   4535.0   \n",
       "4                                   4535.0   \n",
       "...                                    ...   \n",
       "519070                             55225.0   \n",
       "519071                         100505741.0   \n",
       "519072                         100505741.0   \n",
       "519073                         100505741.0   \n",
       "519074                         100505741.0   \n",
       "\n",
       "        BioGRID Interaction data, The General Repository for Interaction Datasets ID  \\\n",
       "0                                                110631.0                              \n",
       "1                                                110631.0                              \n",
       "2                                                110631.0                              \n",
       "3                                                110631.0                              \n",
       "4                                                110631.0                              \n",
       "...                                                   ...                              \n",
       "519070                                                NaN                              \n",
       "519071                                                NaN                              \n",
       "519072                                                NaN                              \n",
       "519073                                                NaN                              \n",
       "519074                                                NaN                              \n",
       "\n",
       "       Gene name  \n",
       "0         MT-ND1  \n",
       "1         MT-ND1  \n",
       "2         MT-ND1  \n",
       "3         MT-ND1  \n",
       "4         MT-ND1  \n",
       "...          ...  \n",
       "519070    RAVER2  \n",
       "519071    SPATA1  \n",
       "519072    SPATA1  \n",
       "519073    SPATA1  \n",
       "519074    SPATA1  \n",
       "\n",
       "[519075 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biomart_table"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Preprocessing\n",
    "\n",
    "We have preprocessed the gene expression data by performing normalization, quality control, and mapping gene symbols to standardized identifiers. For PPI data, we filtered interactions based on a confidence score threshold and mapped protein identifiers to gene identifiers.\n",
    "\n",
    "### Normalization Methods\n",
    "\n",
    "- Gene expression data: RPKM/FPKM, TPM, or other normalization methods depending on the data type (microarray or RNA-seq)\n",
    "\n",
    "### Quality Control\n",
    "\n",
    "- Removal of low-quality samples or genes\n",
    "- Removal of batch effects using methods like ComBat\n",
    "\n",
    "### Identifier Mapping\n",
    "\n",
    "- Mapping gene symbols to Entrez Gene IDs or Ensembl Gene IDs\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exploratory Data Analysis\n",
    "\n",
    "In this section, we present a series of plots and visualizations that help us understand the distribution and properties of our data.\n",
    "\n",
    "### Gene Expression Data\n",
    "\n",
    "- Sample distribution across datasets, tissue types, and organisms\n",
    "- Heatmap of gene expression values\n",
    "- Principal Component Analysis (PCA) or t-SNE plots\n",
    "\n",
    "### Protein-Protein Interaction Data\n",
    "\n",
    "- Distribution of interaction confidence scores\n",
    "- Degree distribution of the PPI network\n",
    "- Distribution of network-based features (e.g., node degree, betweenness centrality, clustering coefficient)\n",
    "\n",
    "This notebook provides an overview of the data used in our study. The next steps involve generating features, selecting relevant genes and network features, and training machine learning models for biomarker identification."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "atd",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
